{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0f155b95-9f5f-43e9-ae06-94151b69fb01",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import time\n",
    "import pickle\n",
    "import os\n",
    "import scipy.stats as stats\n",
    "import sigfig as sf\n",
    "\n",
    "import helper_functions as hf\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d017344a-cd7e-4664-9f9d-3f3d42517aa6",
   "metadata": {},
   "outputs": [],
   "source": [
    "tfCore='NWNNGANA'\n",
    "all8merSeqsPossible=hf.IupacToAllPossibleSequences(tfCore)\n",
    "\n",
    "all8merSeqsPossible+=[hf.revcomp(seq) for seq in all8merSeqsPossible]\n",
    "print(all8merSeqsPossible[:3])\n",
    "\n",
    "tfName='IRF3'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aa6a4d94-b8d4-4ef7-941f-8f95c6dbe73e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def getAll8mers(seq,p,ref,alt):\n",
    "    \n",
    "    altLenAdj=len(ref)-len(alt)\n",
    "      \n",
    "    if p<8:\n",
    "        # print(0)\n",
    "        seq=seq[:p+8]\n",
    "        seqRef=seq[:p]+ref+seq[p+1:]\n",
    "        seqAlt=seq[:p]+alt+seq[p+1:]\n",
    "        \n",
    "#         print(seq)\n",
    "#         print(seqRef)\n",
    "#         print(seqAlt)\n",
    "    \n",
    "    elif altLenAdj==0:\n",
    "        # print(1)\n",
    "        seq=seq[p-7:p+8]    \n",
    "        seqRef=seq[:7]+ref+seq[7+1:]\n",
    "        seqAlt=seq[:7]+alt+seq[7+1:]\n",
    "        \n",
    "        # print(seq)\n",
    "        # print(seqRef)\n",
    "        # print(seqAlt)\n",
    "        \n",
    "    elif altLenAdj>0:\n",
    "        # print(2)\n",
    "        seq=seq[p-7:p+8+altLenAdj]    \n",
    "        seqRef=seq[:7]+ref+seq[7+1:-altLenAdj]\n",
    "        seqAlt=seq[:7]+alt+seq[7+1:]\n",
    "    else:\n",
    "        # print(3)\n",
    "        raise ValueError('ref/alt not allowed yet')\n",
    "    \n",
    "    # if len(seqRef)!=len(seqAlt):\n",
    "        # raise ValueError('ref/alt seqs not dovetailed')\n",
    "        \n",
    "    for i,end in enumerate(range(7,len(seq)-altLenAdj)):\n",
    "        # print(4)\n",
    "        start=end-7\n",
    "        last=end+1\n",
    "        kmer8Ref=seqRef[start:last]\n",
    "        kmer8Alt=seqAlt[start:last]\n",
    "            \n",
    "        yield kmer8Ref,kmer8Alt\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "426c577f-d328-42eb-8537-d67a79999a18",
   "metadata": {},
   "outputs": [],
   "source": [
    "def is_seq_tfsite_match(kmer8):\n",
    "    if kmer8 in all8merSeqsPossible: return True\n",
    "    else:                    return False\n",
    "\n",
    "# Set up logic for creating/deleting ets sites\n",
    "BeforeAfter2TfLogic={\n",
    "    (0,0):'Tf Never',\n",
    "    (1,1):'Tf Always',\n",
    "    (0,1):'Tf Created',\n",
    "    (1,0):'Tf Deleted'\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4321a122-032e-4d6f-80d8-55ff0c81c292",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def snv_tf_analysis(chrom,idx1,ref,alt,chr2seq,tf2aff):\n",
    "    \n",
    "    # Ensure data is snp, no indels\n",
    "    prohibitedPolyTypes=['Other']\n",
    "    if len(ref)==1 and len(alt)==1: \n",
    "        polymorphismType='SNV'\n",
    "        \n",
    "    elif len(ref)==1 and len(alt)==0:\n",
    "        polymorphismType='SND'\n",
    "        \n",
    "    else:\n",
    "        polymorphismType='Other'\n",
    "        \n",
    "    if polymorphismType in prohibitedPolyTypes:\n",
    "        return 'Variant Type Not Allowed Yet'\n",
    "        \n",
    "    ######################################\n",
    "    # Verify ref/alt matches genomic bp\n",
    "    ######################################\n",
    "    \n",
    "    # if 'chr' not in chrom: chrom=f'chr{chrom}'\n",
    "    \n",
    "    # Check if pos in genome is same as declared REF\n",
    "    idx0=int(idx1)#-1\n",
    "    snvNtInGenome=chr2seq[chrom][idx0]\n",
    "    if ref==snvNtInGenome:    \n",
    "        checkSnv='REF'\n",
    "    elif alt==snvNtInGenome:  \n",
    "        checkSnv='ALT'\n",
    "    else:                     \n",
    "        return f'Genomic bp does not match expectation: \\n  expected ref/alt = ({ref}/{alt})\\n  observed = {snvNtInGenome}'\n",
    "    \n",
    "    \n",
    "    ######################################\n",
    "    # Assess if SNV is in ETS\n",
    "    ######################################\n",
    "    \n",
    "    tfChanged=False\n",
    "    for kmer8,kmer8mut in getAll8mers(chr2seq[chrom],idx0,ref,alt):\n",
    "        # print(kmer8,kmer8mut)\n",
    "\n",
    "        ba=(is_seq_tfsite_match(kmer8),is_seq_tfsite_match(kmer8mut)) # before after logic\n",
    "        tfLogic=BeforeAfter2TfLogic[ba]\n",
    "\n",
    "        # If ets-changing snp found, write it out\n",
    "        if tfLogic=='Tf Always': \n",
    "            tfChanged=True\n",
    "            refaff=tf2aff[kmer8]\n",
    "            altaff=tf2aff[kmer8mut]\n",
    "            \n",
    "            return tfLogic,refaff,altaff,altaff/refaff,kmer8,kmer8mut\n",
    "        \n",
    "        elif tfLogic=='Tf Created':\n",
    "            return tfLogic,kmer8,kmer8mut\n",
    "            \n",
    "        elif tfLogic=='Tf Deleted':\n",
    "            return tfLogic,kmer8,kmer8mut\n",
    "\n",
    "    if tfChanged==False:\n",
    "        return 'Variant not in Tf',kmer8,kmer8mut\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6e19887a-63bf-40ba-b1c1-8129df745feb",
   "metadata": {},
   "source": [
    "# Load ref data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "23208bbb-bf34-4459-8158-fed41295f4f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# IRF\n",
    "ref  =f'preprocessed-data/parsed_Irf3_3985.1_v1_contig8mers_core=NWNNGANA_max=AAAGGAAA.txt'\n",
    "\n",
    "Seq2TfAff  = hf.loadAff(ref)\n",
    "hf.dprint(Seq2TfAff)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "276fe0aa-4a69-42e4-a58f-1da9f81c5010",
   "metadata": {},
   "outputs": [],
   "source": [
    "# hg38\n",
    "ref='preprocessed-data/IFN-genome.fa'\n",
    "chr2seq=hf.faLoadGenome(ref)\n",
    "chr2seq"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0d8b52f4-6577-4367-9934-e58188a9932c",
   "metadata": {},
   "source": [
    "# Load MPRA data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4082b2ab-fe19-4e47-bf26-e302880d3a37",
   "metadata": {},
   "outputs": [],
   "source": [
    "fn='preprocessed-data/IFNB__for-input.tsv'\n",
    "df=pd.read_csv(fn,sep='\\t')\n",
    "df.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dee3d691-a76c-4bce-b159-cc3b30e3c574",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "        \n",
    "\n",
    "# elementList=[]\n",
    "variantTypeList=[]\n",
    "refAffList=[]\n",
    "altAffList=[]\n",
    "fcList=[]\n",
    "ref8merlist=[]\n",
    "alt8merlist=[]\n",
    "\n",
    "for idx in df.index:\n",
    "    \n",
    "    chrom,pos,ref,alt=[df.at[idx,ci] for ci in ['chrom','pos','ref','alt']]\n",
    "\n",
    "    if alt=='-': alt=''\n",
    "    resultList=snv_tf_analysis(chrom,pos,ref,alt,chr2seq,Seq2TfAff)\n",
    "    result=resultList[0]\n",
    "    \n",
    "    if result=='Variant not in Tf':\n",
    "        result,kmer8,kmer8mut=resultList\n",
    "        variantTypeList.append('Not In Tf')\n",
    "        refAffList.append(np.NaN)\n",
    "        altAffList.append(np.NaN)\n",
    "        fcList.append(np.NaN)\n",
    "        ref8merlist.append(kmer8)\n",
    "        alt8merlist.append(kmer8mut)\n",
    "\n",
    "    elif result=='Tf Created':\n",
    "        result,kmer8,kmer8mut=resultList\n",
    "        variantTypeList.append('Tf Created')\n",
    "        refAffList.append(np.NaN)\n",
    "        altAffList.append(np.NaN)\n",
    "        fcList.append(np.NaN)\n",
    "        ref8merlist.append(kmer8)\n",
    "        alt8merlist.append(kmer8mut)\n",
    "\n",
    "    elif result=='Tf Deleted':\n",
    "        result,kmer8,kmer8mut=resultList\n",
    "        variantTypeList.append('Tf Deleted')\n",
    "        refAffList.append(np.NaN)\n",
    "        altAffList.append(np.NaN)\n",
    "        fcList.append(np.NaN)\n",
    "        ref8merlist.append(kmer8)\n",
    "        alt8merlist.append(kmer8mut)\n",
    "        \n",
    "    else:\n",
    "        variantType,refAff,altAff,fc,kmer8ref,kmer8alt=resultList\n",
    "        variantTypeList.append(variantType)\n",
    "        refAffList.append(refAff)\n",
    "        altAffList.append(altAff)\n",
    "        fcList.append(fc)\n",
    "        ref8merlist.append(kmer8ref)\n",
    "        alt8merlist.append(kmer8alt)\n",
    "        \n",
    "        \n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9f17a2df-73fb-4998-9e2d-6d9afad1a331",
   "metadata": {},
   "outputs": [],
   "source": [
    "df['variant-type']=variantTypeList\n",
    "df['aff-ref']=refAffList\n",
    "df['aff-alt']=altAffList\n",
    "df['aff-fc']=fcList\n",
    "df['kmer-ref']=ref8merlist\n",
    "df['kmer-alt']=alt8merlist\n",
    "\n",
    "df['-log(p)']=df['pvalue'].apply(lambda p: -np.log10(p))\n",
    "df['-log(p)*direction']=df.apply(lambda r: r['-log(p)'] if r['effect']>=1 else -r['-log(p)'],axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4d21892d-c7b2-4705-a0ac-625537f46f68",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.loc[df['variant-type']!=False,:].head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6422b263-8ca8-4f28-b5c0-b05c37ac9681",
   "metadata": {},
   "outputs": [],
   "source": [
    "df['variant-type'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d63b00e2-cc93-4b4a-9307-21314255a27f",
   "metadata": {},
   "source": [
    "# Which sites are used"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c0ccbc1e-6337-4357-8a02-8c030f045165",
   "metadata": {},
   "outputs": [],
   "source": [
    "siteDetected=df['kmer-ref']!=False\n",
    "df.loc[siteDetected,'pos'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "59a353ea-6c32-4bcb-9924-804dd769941f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# IFNB\n",
    "#               1         2         3         4         5         6         7         8\n",
    "#     012345678901234567890123456789012345678901234567890123456789012345678901234567890123456\n",
    "# en='TAGAAACTACTAAAATGTAAATGACATAGGAAAACTGAAAGGGAGAAGTGAAAGTGGGAAATTCCTCTGAATAGAGAGAGGACCATC'\n",
    "# IRF3                         25           38\n",
    "# IRF7                                32           45\n",
    "\n",
    "# all bp \n",
    "# irf3=[25,26,27,28,29,30,31,32,      38,39,40,41,42,43,44,45]\n",
    "# irf7=[32,33,34,35,36,37,38,39,      45,46,47,48,49,50,51,52]\n",
    "\n",
    "#25-32; 32-39; 38-45; 45-52\n",
    "\n",
    "# remove overlapping binding sites for which snvs may cause multiple effects\n",
    "irf3=[25,26,27,28,29,30,31,                     40,41,42,43,44]\n",
    "irf7=[                        33,34,35,36,37,                    46,47,48,49,50,51,52]\n",
    "\n",
    "olapTfbsPos=[32,38,39,45]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c6d014f1-6a8d-4a9c-8da5-7552f477dc9b",
   "metadata": {},
   "outputs": [],
   "source": [
    "tfbsPosList=list(irf3+irf7)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ac1a493d-7976-43ed-abf0-cd076a6b6ba0",
   "metadata": {},
   "outputs": [],
   "source": [
    "df['pos-of-tfbs']=df.pos.apply(lambda pos: pos in tfbsPosList)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5b4b38b7-ee7c-403f-8f41-cbb1ec1f2395",
   "metadata": {},
   "outputs": [],
   "source": [
    "# remove begiinning and end of enhancer which dont have canonical tfbs\n",
    "excludePos=list(range(0,16+1)) + list(range(69,86+1)) \n",
    "\n",
    "df['part-of-flank']=df.pos.apply(lambda pos: pos in excludePos)\n",
    "df=df.loc[df['part-of-flank']==False,:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fdab4cb4-0722-4d8e-a1af-9a220ff21145",
   "metadata": {},
   "outputs": [],
   "source": [
    "x=plt.hist(Seq2TfAff.values(),bins=100)\n",
    "plt.xlim(0,1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a8dbb992-7a5e-4577-8a8c-3fc761066465",
   "metadata": {},
   "outputs": [],
   "source": [
    "pvalCutoff=.05\n",
    "\n",
    "deNovoThresh=.25\n",
    "\n",
    "fcOptCutoff=1.5\n",
    "fcDecCutoff=sf.round(1/fcOptCutoff,3)\n",
    "\n",
    "fcNoChangeCutoff_Inc=1.25\n",
    "fcNoChangeCutoff_Dec=sf.round(1/fcNoChangeCutoff_Inc,3)\n",
    "\n",
    "logSigCutoff=abs(np.log10(pvalCutoff))\n",
    "\n",
    "predDF=df.copy(deep=True)\n",
    "\n",
    "# assign prediction labels\n",
    "bins=[-np.inf,-logSigCutoff,logSigCutoff,np.inf]\n",
    "predDF['classification']=pd.cut(predDF['-log(p)*direction'],bins=bins,labels=['Sig Decrease','N.S.','Sig Increase'])\n",
    "\n",
    "# assign predictions\n",
    "tfLabelsList=[]\n",
    "\n",
    "# classes in which you only look at irf3 binding sites\n",
    "onlyLookAtTfbsPos=[\n",
    "    f'no-change-aff\\n<={fcNoChangeCutoff_Inc}\\n>={fcNoChangeCutoff_Dec}',\n",
    "    f'dramatic-increase\\n>={fcOptCutoff}'\n",
    "]\n",
    "\n",
    "for posInTfbs,fc,classification,kmer,kmerMut in hf.zipdf(predDF,['pos-of-tfbs','aff-fc','variant-type','kmer-ref','kmer-alt']):\n",
    "\n",
    "    # not in tf\n",
    "    if classification=='Not In Tf':\n",
    "        anno='not-in-tf'\n",
    "\n",
    "    # delet tf\n",
    "    elif classification=='Tf Deleted':\n",
    "        anno='del-tf'\n",
    "\n",
    "    elif classification=='Tf Created':\n",
    "\n",
    "        denovoaff=Seq2TfAff[kmerMut]\n",
    "\n",
    "        if denovoaff<deNovoThresh:\n",
    "            anno=f'de-novo-tf-low-aff\\n<{deNovoThresh}'            \n",
    "\n",
    "        elif denovoaff>=deNovoThresh:\n",
    "            anno=f'de-novo-tf-high-aff\\n>={deNovoThresh}'\n",
    "\n",
    "    #  dramatic increase\n",
    "    elif fc>=fcOptCutoff and posInTfbs==True:\n",
    "        anno=f'dramatic-increase\\n>={fcOptCutoff}'\n",
    "\n",
    "    #  dramatic decrease\n",
    "    elif fc<=fcDecCutoff and posInTfbs==True:\n",
    "        # print(fc,classification)\n",
    "        anno=f'dramatic-decrease\\n<={fcDecCutoff}'\n",
    "\n",
    "    # no change\n",
    "    elif fc<=fcNoChangeCutoff_Inc and fc>=fcNoChangeCutoff_Dec and posInTfbs==True:\n",
    "        anno=f'no-change-aff\\n<={fcNoChangeCutoff_Inc}\\n>={fcNoChangeCutoff_Dec}'\n",
    "\n",
    "    elif posInTfbs==False:\n",
    "        anno='not-in-tf'\n",
    "        \n",
    "    else:\n",
    "        # print(fc,classification,kmer,kmerMut)\n",
    "        anno='subtle-change'\n",
    "\n",
    "    tfLabelsList.append(anno)\n",
    "\n",
    "predDF['tf-class']=tfLabelsList\n",
    "\n",
    "line_out=hf.write_row(['snv-group','total','effect-ns','effect-gof','effect-lof'])\n",
    "\n",
    "for classi in predDF['tf-class'].unique():\n",
    "    nClassi=predDF['tf-class'].value_counts()[classi]\n",
    "    \n",
    "    if classi in onlyLookAtTfbsPos:\n",
    "        ns,gof,lof=[predDF.loc[(predDF['tf-class']==classi) & (predDF['pos-of-tfbs']==True),'classification'].value_counts()[ci] for ci in ['N.S.','Sig Increase','Sig Decrease']]\n",
    "    else:\n",
    "        ns,gof,lof=[predDF.loc[(predDF['tf-class']==classi)                                ,'classification'].value_counts()[ci] for ci in ['N.S.','Sig Increase','Sig Decrease']]\n",
    "\n",
    "    line_out+=hf.write_row([classi,nClassi,ns,gof,lof])\n",
    "\n",
    "print(line_out)\n",
    "with open(f'FOR-NATURE-STATS-PREDICTIONS-{tfName}__fdr.tsv','w') as f: f.write(line_out)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "35ebd027-9c59-4f07-9aa9-6d37c0aae3dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "altHypothesis='less'\n",
    "violin=False\n",
    "boxplot=True\n",
    "\n",
    "datalabels_subset_v2=[\n",
    "[f'not-in-tf',f'del-tf',f'dramatic-decrease\\n<={fcDecCutoff}',f'de-novo-tf-low-aff\\n<{deNovoThresh}',f'de-novo-tf-high-aff\\n>={deNovoThresh}','subtle-change'],\n",
    "[f'no-change-aff\\n<={fcNoChangeCutoff_Inc}\\n>={fcNoChangeCutoff_Dec}'],\n",
    "[f'dramatic-increase\\n>={fcOptCutoff}'],\n",
    "]\n",
    "\n",
    "data2color={\n",
    "(f'no-change-aff\\n<={fcNoChangeCutoff_Inc}\\n>={fcNoChangeCutoff_Dec}',):'lightgrey',\n",
    "(f'dramatic-increase\\n>={fcOptCutoff}',):'forestgreen',\n",
    "(f'not-in-tf',f'del-tf',f'dramatic-decrease\\n<={fcDecCutoff}',f'de-novo-tf-low-aff\\n<{deNovoThresh}',f'de-novo-tf-high-aff\\n>={deNovoThresh}','subtle-change'):'dimgrey'}\n",
    "\n",
    "\n",
    "plotDF=predDF.copy(deep=True)\n",
    "\n",
    "data=[]\n",
    "for di in datalabels_subset_v2:\n",
    "    datai=[]\n",
    "    for dj in di:\n",
    "        if dj in onlyLookAtTfbsPos:\n",
    "            datai+=plotDF.loc[(plotDF['tf-class']==dj) & (plotDF['pos-of-tfbs']==True),'-log(p)*direction'].tolist()\n",
    "        else: \n",
    "            datai+=plotDF.loc[(plotDF['tf-class']==dj),                                '-log(p)*direction'].tolist()\n",
    "    data.append(datai)\n",
    "\n",
    "line_out=hf.write_row(['group1','group2','p'])\n",
    "\n",
    "############################################################################\n",
    "# Stats\n",
    "############################################################################\n",
    "\n",
    "dCtrlNameList=datalabels_subset_v2[0]\n",
    "dCtrl=plotDF.loc[(plotDF['tf-class'].isin(dCtrlNameList)),'-log(p)*direction']\n",
    "\n",
    "for diNameList in datalabels_subset_v2[1:]:\n",
    "\n",
    "    di=plotDF.loc[(plotDF['tf-class'].isin(diNameList)) & (plotDF['pos-of-tfbs']==True),'-log(p)*direction']\n",
    "\n",
    "    try:\n",
    "        print('dCtrl len',len(dCtrl))\n",
    "        print('di len',len(di))\n",
    "        _ , p =stats.mannwhitneyu(dCtrl,di,alternative=altHypothesis)\n",
    "    except ValueError:\n",
    "        pass\n",
    "\n",
    "\n",
    "    line_out+=hf.write_row(['+'.join(dCtrlNameList).replace('\\n',' '),'+'.join(diNameList).replace('\\n',' '),sf.round(p,4)])\n",
    "\n",
    "print(line_out)\n",
    "with open(f'FOR-NATURE__violin-stats__{tfName}.tsv','w') as f: f.write(line_out)\n",
    "\n",
    "############################################################################\n",
    "# plot\n",
    "############################################################################\n",
    "\n",
    "fig,ax=hf.quickfig(2*len(datalabels_subset_v2),4,dpi=300)\n",
    "\n",
    "palette=[data2color[tuple(di)] for di in datalabels_subset_v2]\n",
    "\n",
    "if violin:\n",
    "    sns.violinplot(data=data,color='white',inner=None,cut=0,ax=ax,width=1.1,scale='area')\n",
    "\n",
    "if boxplot:\n",
    "    sns.boxplot(data=data,color='white',ax=ax,fliersize=0)#,palette=palette)\n",
    "\n",
    "lc=0\n",
    "for xi,di in enumerate(data):\n",
    "    for lc, yi in enumerate(di):\n",
    "        if lc>1000: break\n",
    "        xjittermax=.08\n",
    "        xijitter=np.random.uniform(-xjittermax,xjittermax)\n",
    "        if yi>=logSigCutoff:            color='lightgreen'\n",
    "        elif yi<=-logSigCutoff:          color='tomato'\n",
    "        else:                           color='lightgrey'\n",
    "\n",
    "        ax.scatter(xijitter+xi,yi, color=color,s=8,zorder=100,alpha=1)#,cut=0) \n",
    "\n",
    "ax.spines['top'].set_visible(False)\n",
    "ax.spines['right'].set_visible(False)\n",
    "\n",
    "axticklabels=['\\n'.join(label) for label in datalabels_subset_v2]\n",
    "ax.set_xticklabels([f'{label}\\nN={len(points):,}' for label,points in zip(axticklabels,data)])\n",
    "ax.set_ylabel('-log(bh_fdr)*direction')\n",
    "\n",
    "ax.axhline(-np.log10(.05),ls='--',color='green',alpha=.2)\n",
    "ax.axhline(np.log10(.05),ls='--',color='red',alpha=.2)\n",
    "ax.axhline(np.log10(1),ls='--',color='black',alpha=.2)\n",
    "\n",
    "ax.set_title(f'{tfName}\\n{altHypothesis}')\n",
    "\n",
    "# plt.show()\n",
    "plt.savefig(\"figure-5c.svg\", format=\"svg\")\n",
    "  \n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8b269d02-7eba-4f8e-ad5c-ee9f4504dde9",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
