{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "2d7471c1-c049-4157-9e6b-95c22fc5cc4f",
   "metadata": {},
   "source": [
    "# Method to analyze whether SNV in ETS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "36b7c7dc-07eb-4eab-8e13-3b2f84c90aac",
   "metadata": {},
   "outputs": [],
   "source": [
    "tfbsType='ETS'\n",
    "tfName='ETS'\n",
    "\n",
    "expname='Tewhey'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "556d2e2c-c4b1-41c2-9918-302df1d9c879",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import snv_ets_analysis as se\n",
    "\n",
    "import helper_functions as hf\n",
    "\n",
    "import scipy.stats as stats\n",
    "from operator import itemgetter\n",
    "import copy\n",
    "import sigfig as sf\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import time\n",
    "import pickle\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "326356d2-aea2-4fc7-abd4-09ad6145c004",
   "metadata": {},
   "outputs": [],
   "source": [
    "def getAll8mers(seq,p,ref,alt):\n",
    "    \n",
    "    altLenAdj=len(ref)-len(alt)\n",
    "        \n",
    "    if altLenAdj==0:\n",
    "        seq=seq[p-7:p+8]    \n",
    "        seqRef=seq[:7]+ref+seq[7+1:]\n",
    "        seqAlt=seq[:7]+alt+seq[7+1:]\n",
    "        \n",
    "    elif altLenAdj>0:\n",
    "        seq=seq[p-7:p+8+altLenAdj]    \n",
    "        seqRef=seq[:7]+ref+seq[7+1:-altLenAdj]\n",
    "        seqAlt=seq[:7]+alt+seq[7+1:]\n",
    "    else:\n",
    "        raise ValueError('ref/alt not allowed yet')\n",
    "    \n",
    "    # if len(seqRef)!=len(seqAlt):\n",
    "        # raise ValueError('ref/alt seqs not dovetailed')\n",
    "        \n",
    "    for i,end in enumerate(range(7,len(seq)-altLenAdj)):\n",
    "        start=end-7\n",
    "        last=end+1\n",
    "        kmer8Ref=seqRef[start:last]\n",
    "        kmer8Alt=seqAlt[start:last]\n",
    "            \n",
    "        yield kmer8Ref,kmer8Alt\n",
    "                    #.  012345678901234567890\n",
    "            \n",
    "# for i in getAll8mers(chr2seq['chr7'], 156791183, 'T', ''):\n",
    "    # print(i)\n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "426c577f-d328-42eb-8537-d67a79999a18",
   "metadata": {},
   "outputs": [],
   "source": [
    "def ets(kmer4):\n",
    "    if kmer4 in ['GGAA','GGAT','TTCC','ATCC']: return True\n",
    "    else:                                      return False\n",
    "\n",
    "# Set up logic for creating/deleting ets sites\n",
    "BeforeAfter2EtsLogic={\n",
    "    (0,0):'Ets Never',\n",
    "    (1,1):'Ets Always',\n",
    "    (0,1):'Ets Created',\n",
    "    (1,0):'Ets Deleted'\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6a028beb-3e66-4c5f-a718-c2f1e9690509",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def snv_ets_analysis(chrom,idx1,ref,alt,chr2seq,tf2aff):\n",
    "    \n",
    "    # Ensure data is snp, no indels\n",
    "    prohibitedPolyTypes=['Other']\n",
    "    if len(ref)==1 and len(alt)==1: \n",
    "        polymorphismType='SNV'\n",
    "        \n",
    "    elif len(ref)==1 and len(alt)==0:\n",
    "        polymorphismType='SND'\n",
    "        \n",
    "    else:\n",
    "        polymorphismType='Other'\n",
    "        \n",
    "    if polymorphismType in prohibitedPolyTypes:\n",
    "        return 'Variant Type Not Allowed Yet'\n",
    "        \n",
    "    ######################################\n",
    "    # Verify ref/alt matches genomic bp\n",
    "    ######################################\n",
    "    \n",
    "    if 'chr' not in chrom: chrom=f'chr{chrom}'\n",
    "    \n",
    "    # Check if pos in genome is same as declared REF\n",
    "    idx0=int(idx1)-1\n",
    "    snvNtInGenome=chr2seq[chrom][idx0]\n",
    "    if ref==snvNtInGenome:    \n",
    "        checkSnv='REF'\n",
    "    elif alt==snvNtInGenome:  \n",
    "        checkSnv='ALT'\n",
    "    else:                     \n",
    "        return f'Genomic bp does not match expectation: \\n  expected ref/alt = ({ref}/{alt})\\n  observed = {snvNtInGenome}'\n",
    "    \n",
    "    \n",
    "    ######################################\n",
    "    # Assess if SNV is in ETS\n",
    "    ######################################\n",
    "    \n",
    "    etsChanged=False\n",
    "    for kmer8,kmer8mut in getAll8mers(chr2seq[chrom],idx0,ref,alt):\n",
    "        core   =kmer8[2:6]\n",
    "        coremut=kmer8mut[2:6]\n",
    "\n",
    "        ba=(ets(core),ets(coremut)) # before after logic\n",
    "        etsLogic=BeforeAfter2EtsLogic[ba]\n",
    "\n",
    "        # If ets-changing snp found, write it out\n",
    "        if etsLogic=='Ets Always': \n",
    "            etsChanged=True\n",
    "            refaff=tf2aff[kmer8]\n",
    "            altaff=tf2aff[kmer8mut]\n",
    "            \n",
    "            return etsLogic,refaff,altaff,altaff/refaff,kmer8,kmer8mut\n",
    "        \n",
    "        elif etsLogic=='Ets Created':\n",
    "            return etsLogic,kmer8,kmer8mut\n",
    "            \n",
    "        elif etsLogic=='Ets Deleted':\n",
    "            return etsLogic,kmer8,kmer8mut\n",
    "\n",
    "    if etsChanged==False:\n",
    "        return 'Variant not in ETS',kmer8,kmer8mut\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6e19887a-63bf-40ba-b1c1-8129df745feb",
   "metadata": {},
   "source": [
    "# Load ref data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2ea6fbb1-3d7b-4be4-b725-9c0ac193f93c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Ets1\n",
    "ref  =f'preprocessed-data/parsed_Ets1_8mers.txt'              \n",
    "Seq2EtsAff  = hf.loadAff(ref)\n",
    "hf.dprint(Seq2EtsAff)\n",
    "\n",
    "\n",
    "# hg38\n",
    "ref='/Users/joe/code/ref/genomes/human/hg19/online-only/hg19.fa'\n",
    "chr2seq=hf.faLoadGenome(ref)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0d8b52f4-6577-4367-9934-e58188a9932c",
   "metadata": {},
   "source": [
    "# Load MPRA data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4082b2ab-fe19-4e47-bf26-e302880d3a37",
   "metadata": {},
   "outputs": [],
   "source": [
    "fn='preprocessed-data/tewhey-dataframe-input__fdr__withRC.tsv'\n",
    "df=pd.read_csv(fn,sep='\\t')\n",
    "df.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "01eb846c-9e60-4415-a29d-b4a39940cf85",
   "metadata": {},
   "outputs": [],
   "source": [
    "len(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c3ce6a89-cf82-462b-8eb6-48c9688fa3c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "## New one which grabs 8mer ref and  8mer alt  regardlesso f snv  type\n",
    "\n",
    "variantTypeList=[]\n",
    "refAffList=[]\n",
    "altAffList=[]\n",
    "fcList=[]\n",
    "ref8merlist=[]\n",
    "alt8merlist=[]\n",
    "\n",
    "for idx in df.index:\n",
    "    \n",
    "    chrom,pos,ref,alt=[df.at[idx,ci] for ci in ['chrom','pos','ref','alt']]\n",
    "\n",
    "    if alt=='-': alt=''\n",
    "    resultList=snv_ets_analysis(chrom,pos,ref,alt,chr2seq,Seq2EtsAff)\n",
    "    result=resultList[0]\n",
    "    \n",
    "    if result=='Variant not in ETS':\n",
    "        result,kmer8,kmer8mut=resultList\n",
    "        variantTypeList.append('Not In Ets')\n",
    "        refAffList.append(np.NaN)\n",
    "        altAffList.append(np.NaN)\n",
    "        fcList.append(np.NaN)\n",
    "        ref8merlist.append(kmer8)\n",
    "        alt8merlist.append(kmer8mut)\n",
    "\n",
    "\n",
    "    elif result=='Ets Created':\n",
    "        result,kmer8,kmer8mut=resultList\n",
    "        variantTypeList.append('Ets Created')\n",
    "        refAffList.append(np.NaN)\n",
    "        altAffList.append(np.NaN)\n",
    "        fcList.append(np.NaN)\n",
    "        ref8merlist.append(kmer8)\n",
    "        alt8merlist.append(kmer8mut)\n",
    "\n",
    "\n",
    "    elif result=='Ets Deleted':\n",
    "        result,kmer8,kmer8mut=resultList\n",
    "        variantTypeList.append('Ets Deleted')\n",
    "        refAffList.append(np.NaN)\n",
    "        altAffList.append(np.NaN)\n",
    "        fcList.append(np.NaN)\n",
    "        ref8merlist.append(kmer8)\n",
    "        alt8merlist.append(kmer8mut)\n",
    "\n",
    "        \n",
    "    else:\n",
    "        variantType,refAff,altAff,fc,kmer8ref,kmer8alt=resultList\n",
    "        variantTypeList.append(variantType)\n",
    "        refAffList.append(refAff)\n",
    "        altAffList.append(altAff)\n",
    "        fcList.append(fc)\n",
    "        ref8merlist.append(kmer8ref)\n",
    "        alt8merlist.append(kmer8alt)\n",
    "\n",
    "        \n",
    "        \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "98d96818-7200-4207-a262-2b99f7071f04",
   "metadata": {},
   "outputs": [],
   "source": [
    "df['variant-type']=variantTypeList\n",
    "df['aff-ref']=refAffList\n",
    "df['aff-alt']=altAffList\n",
    "df['aff-fc']=fcList\n",
    "df['kmer-ref']=ref8merlist\n",
    "df['kmer-alt']=alt8merlist\n",
    "\n",
    "df['aff-delta']=df.apply(lambda row: row['aff-alt']-row['aff-ref'],axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "364f8011-616b-463a-892a-84aa3d903029",
   "metadata": {},
   "outputs": [],
   "source": [
    "df['variant-type'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "39cc9ff8-5d0c-4527-9234-b749f658ff85",
   "metadata": {},
   "outputs": [],
   "source": [
    "df['-log(p)']=df['log-p'].apply(lambda logp: logp)\n",
    "df['-log(p)*direction']=df.apply(lambda r: r['-log(p)'] if r['effect']>=0 else -r['-log(p)'],axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4d21892d-c7b2-4705-a0ac-625537f46f68",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.loc[df['variant-type']!=False,:].head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6422b263-8ca8-4f28-b5c0-b05c37ac9681",
   "metadata": {},
   "outputs": [],
   "source": [
    "df['variant-type'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "730437ff-5bd3-4ab1-8199-ba3bc8bdad07",
   "metadata": {},
   "outputs": [],
   "source": [
    "df['chrompos']=df['chrom']+df['pos'].apply(str)\n",
    "print(len(df['chrompos']))\n",
    "print(len(df['chrompos'].unique()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "250af1b9-2c43-4cbb-b0f1-ea35988034e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# some rsids are present twice. Prioritize keeping the tested variant which is significant. If significant in both direction (gof and lof), then do not use\n",
    "\n",
    "duplicateChrompos  = [cp for cp in df['chrompos'].unique() if df['chrompos'].tolist().count(cp)>1]\n",
    "len(duplicateChrompos)\n",
    "\n",
    "logp5=-np.log10(.05)\n",
    "\n",
    "keepIndices=[]\n",
    "duplicateChromposConsidered=[]\n",
    "\n",
    "nomit=0\n",
    "\n",
    "for idx in df.index:\n",
    "    \n",
    "    cp=df.at[idx,'chrompos']\n",
    "    \n",
    "    # if index not a duplicate, simply keep the index\n",
    "    if cp not in duplicateChrompos:\n",
    "        keepIndices.append(idx)\n",
    "        duplicateChromposConsidered.append(cp)\n",
    "        \n",
    "    # if you've already analyzed this cp, then skip\n",
    "    elif cp in duplicateChromposConsidered:\n",
    "        continue\n",
    "    \n",
    "    # if duplicate and haven't seen, choose one of the dups to keep\n",
    "    else:\n",
    "        r1,r2=df.loc[df['chrompos']==cp,'-log(p)*direction'].tolist()\n",
    "        \n",
    "        r1isgof = r1>logp5\n",
    "        r2isgof = r2>logp5\n",
    "        \n",
    "        r1islof = r1<-logp5\n",
    "        r2islof = r2<-logp5\n",
    "        \n",
    "        r1isns  = (not r1isgof) and (not r1islof)\n",
    "        r2isns  = (not r2isgof) and (not r2islof)\n",
    "        \n",
    "        # if BOTH gof, keep r1\n",
    "        if r1isgof and r2isgof:\n",
    "            idxkeep=df.loc[df['chrompos']==cp,:].index.tolist()[0]\n",
    "            keepIndices.append(idxkeep)\n",
    "            \n",
    "        # if ONE is sig... one is n.s.\n",
    "        #          ... where r1 is significant\n",
    "        elif ((r1isgof or r1islof) and r2isns):\n",
    "            idxkeep=df.loc[(df['chrompos']==cp) & (df['-log(p)*direction']==r1),:].index.tolist()[0]\n",
    "            keepIndices.append(idxkeep)\n",
    "            \n",
    "        #          ... where r2 is significant\n",
    "        elif ((r2isgof or r2islof) and r1isns):\n",
    "            idxkeep=df.loc[(df['chrompos']==cp) & (df['-log(p)*direction']==r2),:].index.tolist()[0]\n",
    "            keepIndices.append(idxkeep)\n",
    "            \n",
    "        # if NEITHER sig, keep r1\n",
    "        elif r1isns and r2isns:\n",
    "            idxkeep=df.loc[df['chrompos']==cp,:].index.tolist()[0]\n",
    "            keepIndices.append(idxkeep)\n",
    "            \n",
    "        # OTHERWISE (one sig gof, other is sig lof), omit\n",
    "        else:\n",
    "            nomit+=1\n",
    "            continue\n",
    "            \n",
    "print('num omitted',nomit)\n",
    "print('num kept',len(keepIndices))\n",
    "\n",
    "df=df.loc[keepIndices,:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "37d5b33e-c121-48e7-a38b-6030a6e49c0b",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(len(df))\n",
    "df=df.drop_duplicates()\n",
    "print(len(df))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "600cd00e-d204-4fc3-942c-1dd2a2d2828b",
   "metadata": {},
   "source": [
    "# Prediction accuracy vs random (for nature)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4fda9015-c8aa-4be2-909d-9d2155455490",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "fcOptCutoff=1.59\n",
    "fcDecCutoff=1/fcOptCutoff\n",
    "\n",
    "pvalCutoff=.05\n",
    "logSigCutoff=abs(np.log10(pvalCutoff))\n",
    "\n",
    "fcNoChangeCutoff_Inc=1.25\n",
    "fcNoChangeCutoff_Dec=sf.round(1/fcNoChangeCutoff_Inc,3)\n",
    "\n",
    "deNovoThresh=.15\n",
    "\n",
    "# get variants to predict on\n",
    "predDF=df.loc[:,:]\n",
    "\n",
    "# assign prediction labels\n",
    "bins=[-np.inf,-logSigCutoff,logSigCutoff,np.inf]\n",
    "predDF['classification']=pd.cut(predDF['-log(p)*direction'],bins=bins,labels=['Sig Decrease','N.S.','Sig Increase'])\n",
    "\n",
    "# assign predictions\n",
    "etsLabels=[]\n",
    "\n",
    "for fc,classification,kmer,kmerMut in hf.zipdf(predDF,['aff-fc','variant-type','kmer-ref','kmer-alt']):\n",
    "\n",
    "    # not in tf\n",
    "    if classification=='Not In Ets':\n",
    "        anno='not-in-tf'\n",
    "\n",
    "    # delet tf\n",
    "    elif classification=='Ets Deleted':\n",
    "        anno='del-tf'\n",
    "\n",
    "    elif classification=='Ets Created':\n",
    "\n",
    "        denovoaff=Seq2EtsAff[kmerMut]\n",
    "\n",
    "        if denovoaff<deNovoThresh:\n",
    "            anno=f'de-novo-tf-low-aff\\n<{deNovoThresh}'            \n",
    "\n",
    "        elif denovoaff>=deNovoThresh:\n",
    "            anno=f'de-novo-tf-high-aff\\n>={deNovoThresh}'\n",
    "\n",
    "    #  dramatic iincrase\n",
    "    elif fc>=fcOptCutoff:\n",
    "        anno=f'dramatic-increase\\n>={fcOptCutoff}'\n",
    "\n",
    "    #  dramatic incrase\n",
    "    elif fc<=fcDecCutoff:\n",
    "        # print(fc,classification)\n",
    "        anno=f'dramatic-decrease\\n<={fcDecCutoff}'\n",
    "\n",
    "    # no increase\n",
    "    elif fc<=fcNoChangeCutoff_Inc and fc>=fcNoChangeCutoff_Dec:\n",
    "        anno=f'no-change-aff\\n<={fcNoChangeCutoff_Inc}\\n>={fcNoChangeCutoff_Dec}'\n",
    "        # print(anno.replace('\\n',' '),fc,classification,kmer,kmerMut)\n",
    "\n",
    "\n",
    "    else:\n",
    "        anno='subtle-change'\n",
    "        # print(anno.replace('\\n',' '),fc,classification,kmer,kmerMut)\n",
    "\n",
    "\n",
    "    etsLabels.append(anno)\n",
    "\n",
    "predDF['ets-class']=etsLabels\n",
    "\n",
    "line_out=hf.write_row(['snv-group','total','effect-ns','effect-gof','effect-lof'])\n",
    "\n",
    "for classi in predDF['ets-class'].unique():\n",
    "    nClassi=predDF['ets-class'].value_counts()[classi]\n",
    "\n",
    "    ns,gof,lof=[predDF.loc[predDF['ets-class']==classi,'classification'].value_counts()[ci] for ci in ['N.S.','Sig Increase','Sig Decrease']]\n",
    "\n",
    "    line_out+=hf.write_row([classi.replace('\\n',' '),nClassi,ns,gof,lof])\n",
    "\n",
    "print(line_out)\n",
    "with open(f'FOR-NATURE-STATS-PREDICTIONS-{expname}-{tfName}__fdr.tsv','w') as f: f.write(line_out)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a847ddbb-7f98-40d2-80f0-404535d5ebea",
   "metadata": {},
   "outputs": [],
   "source": [
    "altHypothesis='less'\n",
    "\n",
    "violin=False\n",
    "boxplot=True\n",
    "\n",
    "datalabels_subset_v2=[\n",
    "[f'not-in-tf',f'del-tf',f'dramatic-decrease\\n<={fcDecCutoff}',f'de-novo-tf-low-aff\\n<{deNovoThresh}',f'de-novo-tf-high-aff\\n>={deNovoThresh}','subtle-change'],\n",
    "[f'no-change-aff\\n<={fcNoChangeCutoff_Inc}\\n>={fcNoChangeCutoff_Dec}'],\n",
    "[f'dramatic-increase\\n>={fcOptCutoff}'],\n",
    "]\n",
    "\n",
    "data2color={\n",
    "(f'no-change-aff\\n<={fcNoChangeCutoff_Inc}\\n>={fcNoChangeCutoff_Dec}',):'lightgrey',\n",
    "(f'dramatic-increase\\n>={fcOptCutoff}',):'forestgreen',\n",
    "(f'not-in-tf',f'del-tf',f'dramatic-decrease\\n<={fcDecCutoff}',f'de-novo-tf-low-aff\\n<{deNovoThresh}',f'de-novo-tf-high-aff\\n>={deNovoThresh}','subtle-change'):'dimgrey'}\n",
    "\n",
    "plotDF=predDF.loc[:,:]\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "data=[]\n",
    "for di in datalabels_subset_v2:\n",
    "    datai=[]\n",
    "    for dj in di:\n",
    "        datai+=plotDF.loc[(plotDF['ets-class']==dj),'-log(p)*direction'].tolist()\n",
    "    data.append(datai)\n",
    "\n",
    "line_out=hf.write_row(['group1','group2','p'])\n",
    "\n",
    "############################################################################\n",
    "# Stats\n",
    "############################################################################\n",
    "\n",
    "dCtrlNameList=datalabels_subset_v2[0]\n",
    "dCtrl=plotDF.loc[(plotDF['ets-class'].isin(dCtrlNameList)),'-log(p)*direction']\n",
    "\n",
    "for diNameList in datalabels_subset_v2[1:]:\n",
    "\n",
    "    di=plotDF.loc[(plotDF['ets-class'].isin(diNameList)),'-log(p)*direction']\n",
    "\n",
    "    try:\n",
    "        _ , p =stats.mannwhitneyu(dCtrl,di,alternative=altHypothesis)\n",
    "    except ValueError:\n",
    "        pass\n",
    "\n",
    "\n",
    "    line_out+=hf.write_row(['+'.join(dCtrlNameList).replace('\\n',' '),'+'.join(diNameList).replace('\\n',' '),sf.round(p,4)])\n",
    "\n",
    "print(line_out)\n",
    "with open(f'FOR-NATURE__violin-stats__{tfName}__{expname}.tsv','w') as f: f.write(line_out)\n",
    "\n",
    "############################################################################\n",
    "# Stats\n",
    "############################################################################\n",
    "\n",
    "fig,ax=hf.quickfig(2*len(datalabels_subset_v2),4,dpi=300)\n",
    "\n",
    "palette=[data2color[tuple(di)] for di in datalabels_subset_v2]\n",
    "\n",
    "if violin:\n",
    "    sns.violinplot(data=data,color='white',inner=None,cut=0,ax=ax,width=1.1,scale='area')\n",
    "\n",
    "if boxplot:\n",
    "    sns.boxplot(data=data,color='white',ax=ax,fliersize=0)#,palette=palette)\n",
    "\n",
    "lc=0\n",
    "for xi,di in enumerate(data):\n",
    "    for lc, yi in enumerate(di):\n",
    "        if lc>1000: break\n",
    "        xjittermax=.08\n",
    "        xijitter=np.random.uniform(-xjittermax,xjittermax)\n",
    "        if yi>=logSigCutoff:            color='lightgreen'\n",
    "        elif yi<=-logSigCutoff:          color='tomato'\n",
    "        else:                           color='lightgrey'\n",
    "\n",
    "        ax.scatter(xijitter+xi,yi, color=color,s=8,zorder=100,alpha=1)#,cut=0) \n",
    "\n",
    "ax.spines['top'].set_visible(False)\n",
    "ax.spines['right'].set_visible(False)\n",
    "\n",
    "axticklabels=['\\n'.join(label) for label in datalabels_subset_v2]\n",
    "ax.set_xticklabels([f'{label}\\nN={len(points):,}' for label,points in zip(axticklabels,data)])\n",
    "ax.set_ylabel('-log(bh_fdr)*direction')\n",
    "\n",
    "ax.axhline(-np.log10(.05),ls='--',color='green',alpha=.2)\n",
    "ax.axhline(np.log10(.05),ls='--',color='red',alpha=.2)\n",
    "ax.axhline(np.log10(1),ls='--',color='black',alpha=.2)\n",
    "\n",
    "ax.set_title(f'{expname}\\n{tfName}\\n{altHypothesis}')\n",
    "\n",
    "\n",
    "# plt.savefig(f'/Users/joe/Desktop/ets-kircher-elements={idList}.png')# ,bbox_inches='tight')\n",
    "plt.show()\n",
    "      \n",
    "\n",
    "plotDF.to_csv(f'NATURE-ALL-DATA_{tfName}_{expname}.tsv',sep='\\t',index=None)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5da6d7ab-b5ee-4d38-9367-2a8bc5aa8cc8",
   "metadata": {},
   "source": [
    "# Adding in eqtl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9e875b7b-72eb-4aa3-b519-3f1d7cf77b57",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "fn=f'preprocessed-data/QTD000110__readyForInput__matteoCorrMethod_keepOnlyTewheyMPRAVariants=True_removeMixedGofLofVariants=False_genomeWideSigAdj=True.tsv'\n",
    "\n",
    "rsid2GeneBetaPvalue={}\n",
    "rsid2EffectAnnotation={}\n",
    "rsidList=[]\n",
    "\n",
    "for _,_,rsid,_,_,beta,pvalue,gene in hf.read_tsv(fn,pc=True,header=False):\n",
    "\n",
    "    beta=float(beta)\n",
    "    pvalue=float(pvalue)\n",
    "\n",
    "    if rsid not in rsid2GeneBetaPvalue:\n",
    "        rsid2GeneBetaPvalue[rsid]=[]\n",
    "    rsid2GeneBetaPvalue[rsid].append((gene,beta,pvalue))\n",
    "\n",
    "    if rsid not in rsid2EffectAnnotation:\n",
    "        rsid2EffectAnnotation[rsid]=set()\n",
    "\n",
    "    if beta<0    and pvalue<.05:   rsid2EffectAnnotation[rsid].add('LOF')\n",
    "    elif beta>0  and pvalue<.05: rsid2EffectAnnotation[rsid].add('GOF')\n",
    "\n",
    "    rsidList.append(rsid)\n",
    "\n",
    "print(len(rsidList),len(set(rsidList)))\n",
    "    \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9ec03cdf-9110-4bd2-949e-bae4e3bad0dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "chrompos2rsid={}\n",
    "rsid2cp={}\n",
    "\n",
    "fn='preprocessed-data//master.geuvadis.annot.parsed.uniq.alleles.reformat.tsv'\n",
    "rsiddf=pd.read_csv(fn,sep='\\t',header=None)\n",
    "rsiddf.columns=['chrom','pos','rsid','ref','alt','strand']\n",
    "rsiddf['chrom']='chr'+rsiddf['chrom'].astype(str)\n",
    "\n",
    "for chrom,pos,rsid in hf.zipdf(rsiddf,['chrom','pos','rsid']):\n",
    "    chrompos=chrom+str(pos)\n",
    "    chrompos2rsid[chrompos]=rsid\n",
    "    rsid2cp[rsid]=(chrom,pos)\n",
    "    \n",
    "# rsiddf.tail()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8b9e8412-42bc-4bbd-ace1-a8cc126075d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "plotDF['rsid']=plotDF['chrompos'].apply(lambda cp: chrompos2rsid[cp])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "12f291ef-6666-4ca3-9c68-76f6fb43d197",
   "metadata": {},
   "outputs": [],
   "source": [
    "plotDF.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b7899c01-4a69-4266-9726-c93b31c1bf66",
   "metadata": {},
   "outputs": [],
   "source": [
    "rsid2targetgene={}\n",
    "rsidUsed=set()\n",
    "data=[]\n",
    "for di in datalabels_subset_v2:\n",
    "    datai=[]\n",
    "    for dj in di:\n",
    "        print(dj)\n",
    "        rsidToInclude=plotDF.loc[(plotDF['ets-class']==dj),'rsid'].tolist()\n",
    "        \n",
    "        dataj=[]\n",
    "        \n",
    "        for rsid in rsidToInclude:\n",
    "\n",
    "            if rsid in rsid2GeneBetaPvalue:\n",
    "                effectList=rsid2GeneBetaPvalue[rsid]\n",
    "                \n",
    "                rsidUsed.add(rsid)\n",
    "                gene,beta,pvalue=sorted(rsid2GeneBetaPvalue[rsid],key=itemgetter(2))[0] # choose most sig variant\n",
    "                \n",
    "                rsid2targetgene[rsid]=gene\n",
    "\n",
    "                if beta>=0: value=-np.log10(pvalue)\n",
    "                if beta<0: value= np.log10(pvalue)\n",
    "                \n",
    "                dataj.append(value)\n",
    "        print()\n",
    "        print('[TOTAL]',len(dataj))\n",
    "        datai.append(dataj)\n",
    "    print()\n",
    "    print('[NAN]',len([i for i in datai if str(i)=='nan']))\n",
    "    print('$$$$$$$$$$$$$$$$$$$$$$$$')\n",
    "    \n",
    "    data.append([i for i in datai if str(i)!='nan'])\n",
    "\n",
    "line_out=hf.write_row(['group1','group2','p'])\n",
    "\n",
    "############################################################################\n",
    "# Stats\n",
    "############################################################################\n",
    "\n",
    "data_plotted=[]\n",
    "dCtrlNameList=datalabels_subset_v2[0]\n",
    "dCtrl=copy.copy(data[0])\n",
    "dCtrl=[i for i in dCtrl if str(i)!='nan']\n",
    "dCtrl=hf.flatten_list(dCtrl)\n",
    "data_plotted.append(dCtrl)\n",
    "\n",
    "for diNameList,di in zip(copy.copy(datalabels_subset_v2[1:]),copy.copy(data[1:])):\n",
    "\n",
    "    di=[i for i in di if str(i)!='nan']\n",
    "    di=hf.flatten_list(di)\n",
    "    data_plotted.append(di)\n",
    "\n",
    "    try:\n",
    "        _ , p =stats.mannwhitneyu(dCtrl,di,alternative=altHypothesis)\n",
    "        p=sf.round(p,4)\n",
    "    except ValueError:\n",
    "        p='valueError'\n",
    "\n",
    "\n",
    "    line_out+=hf.write_row(['+'.join(dCtrlNameList).replace('\\n',' '),'+'.join(diNameList).replace('\\n',' '),p])\n",
    "\n",
    "print(line_out)\n",
    "with open(f'FOR-NATURE__violin-stats__eqtl__{tfName}__{expname}.tsv','w') as f: f.write(line_out)\n",
    "\n",
    "############################################################################\n",
    "# Stats\n",
    "############################################################################\n",
    "\n",
    "fig,ax=hf.quickfig(2*len(datalabels_subset_v2),4,dpi=300)\n",
    "\n",
    "palette=[data2color[tuple(di)] for di in datalabels_subset_v2]\n",
    "\n",
    "if violin:\n",
    "    sns.violinplot(data=data,color='white',inner=None,cut=0,ax=ax,width=1.1,scale='area')\n",
    "\n",
    "if boxplot:\n",
    "    sns.boxplot(data=data_plotted,color='white',ax=ax,fliersize=0)#,palette=palette)\n",
    "\n",
    "lc=0\n",
    "for xi,di in enumerate(data_plotted):\n",
    "    for lc, yi in enumerate(di):\n",
    "        if lc>1000: break\n",
    "        xjittermax=.08\n",
    "        xijitter=np.random.uniform(-xjittermax,xjittermax)\n",
    "        if yi>=logSigCutoff:            color='lightgreen'\n",
    "        elif yi<=-logSigCutoff:          color='tomato'\n",
    "        else:                           color='lightgrey'\n",
    "\n",
    "        ax.scatter(xijitter+xi,yi, color=color,s=8,zorder=100,alpha=1)#,cut=0) \n",
    "\n",
    "ax.spines['top'].set_visible(False)\n",
    "ax.spines['right'].set_visible(False)\n",
    "\n",
    "axticklabels=['\\n'.join(label) for label in datalabels_subset_v2]\n",
    "ax.set_xticklabels([f'{label}\\nN={len(points):,}' for label,points in zip(axticklabels,data_plotted)])\n",
    "ax.set_ylabel('eqtl\\n-log(bh_fdr)*direction')\n",
    "\n",
    "ax.axhline(-np.log10(.05),ls='--',color='green',alpha=.2)\n",
    "ax.axhline(np.log10(.05),ls='--',color='red',alpha=.2)\n",
    "ax.axhline(np.log10(1),ls='--',color='black',alpha=.2)\n",
    "\n",
    "ax.set_title(f'{expname} - Gauv\\n{tfName}\\n{altHypothesis}')\n",
    "ax.set_ylim(-120,175)\n",
    "\n",
    "\n",
    "\n",
    "# plt.savefig(f'/Users/joe/Desktop/ets-kircher-elements={idList}.png')# ,bbox_inches='tight')\n",
    "plt.show()\n",
    "\n",
    "\n",
    "print(len(rsidUsed))\n",
    "plotDF['target-gene']=plotDF.rsid.apply(lambda rsid: rsid2targetgene[rsid] if rsid in rsid2targetgene else np.NaN)\n",
    "plotDF.loc[plotDF.rsid.isin(rsidUsed),:].to_csv(f'OCTOBER-NATURE-ALL-EQTL-DATA_{tfName}_{expname}.tsv',sep='\\t',index=None)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "98b0d112-97c9-4dd7-a653-176c9462b1a9",
   "metadata": {},
   "source": [
    "## Correlation of tewhey mpra with eqtl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1817e0d1-a979-4fcd-8a45-2e31d1cb15a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "optets=plotDF['ets-class']=='dramatic-increase\\n>=1.59'\n",
    "mpragof=plotDF['classification']=='Sig Increase'\n",
    "\n",
    "rsidListOptEtsAndGofMpra=plotDF.loc[optets & mpragof,'rsid'].tolist()\n",
    "rsidListOptEtsAndGofMpra,len(rsidListOptEtsAndGofMpra)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "24a117f8-6eb4-4e1d-84aa-e681a58e42b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "dogdf=plotDF.loc[optets & mpragof,:].copy(deep=True).set_index('rsid')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d1e06459-a97d-473f-8748-443496cc0c6c",
   "metadata": {},
   "outputs": [],
   "source": [
    "lo=hf.getLiftoverObject('hg38','hg19')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cd31c311-ed15-4933-b0ff-ce7782150284",
   "metadata": {},
   "outputs": [],
   "source": [
    "mpra=[]\n",
    "eqtl=[]\n",
    "rsidPlotted=[]\n",
    "\n",
    "rsid2rsidCorrect={'rs111962081':'rs62055948'}\n",
    "\n",
    "eqtlMpraTable={c:[] for c in ['chrom','pos','ref','alt','rsid','aff_fc','mpra_-log(p)*direction','eqtl_-log(p)*direction','eqtl_gene_id']}\n",
    "\n",
    "for rsid in rsidListOptEtsAndGofMpra:\n",
    "    \n",
    "    # make sure effect is one direction for all genes\n",
    "    if rsid in rsid2EffectAnnotation or rsid2rsidCorrect[rsid] in rsid2EffectAnnotation:\n",
    "        \n",
    "        if rsid in rsid2rsidCorrect:\n",
    "            if rsid2rsidCorrect[rsid] in rsid2EffectAnnotation: \n",
    "\n",
    "                rsid2=rsid2rsidCorrect[rsid]\n",
    "        else:\n",
    "            rsid2=rsid\n",
    "        \n",
    "        if len(rsid2EffectAnnotation[rsid2])==1:\n",
    "\n",
    "            effectList=rsid2GeneBetaPvalue[rsid]\n",
    "\n",
    "            for gene,eqtlbeta,eqtlpvalue in effectList:\n",
    "                if eqtlpvalue>.05: continue\n",
    "                if eqtlbeta>0: eqtldir=1\n",
    "                else:          eqtldir=-1\n",
    "                    \n",
    "                mpravalue=dogdf.at[rsid,'-log(p)*direction']\n",
    "                eqtlvalue=-np.log10(eqtlpvalue)*eqtldir\n",
    "                \n",
    "                eqtl.append(eqtlvalue)\n",
    "                mpra.append(mpravalue)\n",
    "                rsidPlotted.append(rsid)\n",
    "                \n",
    "                chrom,pos,ref,alt,aff_fcc = [dogdf.at[rsid,ci] for ci in ['chrom','pos','ref','alt','aff-fc']]\n",
    "                \n",
    "                eqtlMpraTable['chrom'].append(chrom)\n",
    "                eqtlMpraTable['pos'].append(pos)\n",
    "                eqtlMpraTable['ref'].append(ref)\n",
    "                eqtlMpraTable['alt'].append(alt)\n",
    "                \n",
    "                if rsid in rsid2rsidCorrect:\n",
    "                    eqtlMpraTable['rsid'].append(rsid2)\n",
    "                    \n",
    "                else:\n",
    "                    eqtlMpraTable['rsid'].append(rsid)\n",
    "                    \n",
    "                eqtlMpraTable['aff_fc'].append(aff_fcc)\n",
    "                eqtlMpraTable['mpra_-log(p)*direction'].append(mpravalue)\n",
    "                eqtlMpraTable['eqtl_-log(p)*direction'].append(eqtlvalue)\n",
    "                eqtlMpraTable['eqtl_gene_id'].append(gene)\n",
    "                \n",
    "                \n",
    "        elif len(rsid2EffectAnnotation[rsid2])==0:\n",
    "            \n",
    "            mpravalue=dogdf.at[rsid,'-log(p)*direction']\n",
    "            eqtlvalue=0\n",
    "\n",
    "            eqtl.append(eqtlvalue)\n",
    "            mpra.append(mpravalue)\n",
    "            rsidPlotted.append(rsid)\n",
    "\n",
    "            chrom,pos,ref,alt,aff_fcc = [dogdf.at[rsid,ci] for ci in ['chrom','pos','ref','alt','aff-fc']]\n",
    "\n",
    "            eqtlMpraTable['chrom'].append(chrom)\n",
    "            eqtlMpraTable['pos'].append(pos)\n",
    "            eqtlMpraTable['ref'].append(ref)\n",
    "            eqtlMpraTable['alt'].append(alt)\n",
    "            eqtlMpraTable['rsid'].append(rsid)\n",
    "\n",
    "            eqtlMpraTable['aff_fc'].append(aff_fcc)\n",
    "            eqtlMpraTable['mpra_-log(p)*direction'].append(mpravalue)\n",
    "            eqtlMpraTable['eqtl_-log(p)*direction'].append(eqtlvalue)\n",
    "            eqtlMpraTable['eqtl_gene_id'].append(gene)\n",
    "            \n",
    "        # take most sig variant\n",
    "        else:\n",
    "            effectList=rsid2GeneBetaPvalue[rsid2\n",
    "                                          ]\n",
    "            for gene,eqtlbeta,eqtlpvalue in sorted(effectList, key=itemgetter(2)):\n",
    "                \n",
    "                if eqtlpvalue>.05: continue\n",
    "                \n",
    "                if eqtlbeta>0: eqtldir=1\n",
    "                else:          eqtldir=-1\n",
    "                    \n",
    "                mpravalue=dogdf.at[rsid,'-log(p)*direction']\n",
    "                eqtlvalue=-np.log10(eqtlpvalue)*eqtldir\n",
    "                \n",
    "                eqtl.append(eqtlvalue)\n",
    "                mpra.append(mpravalue)\n",
    "                rsidPlotted.append(rsid)\n",
    "                \n",
    "                chrom,pos,ref,alt,aff_fcc = [dogdf.at[rsid,ci] for ci in ['chrom','pos','ref','alt','aff-fc']]\n",
    "                \n",
    "                eqtlMpraTable['chrom'].append(chrom)\n",
    "                eqtlMpraTable['pos'].append(pos)\n",
    "                eqtlMpraTable['ref'].append(ref)\n",
    "                eqtlMpraTable['alt'].append(alt)\n",
    "                \n",
    "                if rsid in rsid2rsidCorrect:\n",
    "\n",
    "                    eqtlMpraTable['rsid'].append(rsid2)\n",
    "                    \n",
    "                else:\n",
    "                    eqtlMpraTable['rsid'].append(rsid)\n",
    "                    \n",
    "                eqtlMpraTable['aff_fc'].append(aff_fcc)\n",
    "                eqtlMpraTable['mpra_-log(p)*direction'].append(mpravalue)\n",
    "                eqtlMpraTable['eqtl_-log(p)*direction'].append(eqtlvalue)\n",
    "                eqtlMpraTable['eqtl_gene_id'].append(gene)\n",
    "                \n",
    "                break\n",
    "                \n",
    "\n",
    "    else:\n",
    "        mpravalue=dogdf.at[rsid,'-log(p)*direction']\n",
    "\n",
    "        \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "36544671-b793-4d8f-8bc7-2a34cfd3398c",
   "metadata": {},
   "outputs": [],
   "source": [
    "outmpreeqtldf=pd.DataFrame(eqtlMpraTable)\n",
    "outmpreeqtldf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "460193c9-05a9-429b-acf4-5594099cccf1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# if rsid-gene sassociation found in more than one dataset (1kg,eur,yri), take the max sig\n",
    "outmpreeqtldf=outmpreeqtldf.groupby(['chrom','pos','eqtl_gene_id']).max().reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3abdb1c7-421e-4233-94d6-87a565736936",
   "metadata": {},
   "outputs": [],
   "source": [
    "outmpreeqtldf.to_csv('FOR-NATURE__ETS-TEWHEY-EQTL-TEWHEY-MPRA.tsv',sep='\\t',index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "50f337e4-d106-4df2-aa17-6ae3d403e5b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig,ax=hf.quickfig(5,5,250)\n",
    "\n",
    "ax.scatter(outmpreeqtldf['mpra_-log(p)*direction'],outmpreeqtldf['eqtl_-log(p)*direction'],color='lightgreen')\n",
    "\n",
    "\n",
    "ax.axhline(-np.log10(.05),ls='--',color='green',alpha=.2)\n",
    "ax.axhline( np.log10(.05),ls='--',color='red',alpha=.2  )\n",
    "ax.axvline(-np.log10(.05),ls='--',color='green',alpha=.2)\n",
    "ax.axvline( np.log10(.05),ls='--',color='red',alpha=.2  )\n",
    "\n",
    "ax.set_xlabel('MPRA -log(p)*direction')\n",
    "ax.set_ylabel('eqtl -log(p)*direction')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
