{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ba4e726f-272a-468d-9d5f-949b43d79b29",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scipy.stats as stats\n",
    "import sigfig as sf\n",
    "import math\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import time\n",
    "import pickle\n",
    "import os\n",
    "\n",
    "import helper_functions as hf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "36b7c7dc-07eb-4eab-8e13-3b2f84c90aac",
   "metadata": {},
   "outputs": [],
   "source": [
    "tfbsType='AP1'\n",
    "tfName='AP1'\n",
    "\n",
    "\n",
    "expname='GEUV'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2ab0691d-d355-4efa-a2b0-6b59d7e89dbc",
   "metadata": {},
   "outputs": [],
   "source": [
    "tfCore='NTKANNMA'\n",
    "\n",
    "tfAllPossible8mers=hf.IupacToAllPossibleSequences(tfCore)\n",
    "tfAllPossible8mers+=[hf.revcomp(seq) for seq in tfAllPossible8mers]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9a2c2dc3-63cd-4fa5-a400-3c9725c181ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "def getAll8mers(seq,p,ref,alt):\n",
    "    \n",
    "    altLenAdj=len(ref)-len(alt)\n",
    "        \n",
    "    if altLenAdj==0:\n",
    "        seq=seq[p-7:p+8]    \n",
    "        seqRef=seq[:7]+ref+seq[7+1:]\n",
    "        seqAlt=seq[:7]+alt+seq[7+1:]\n",
    "        \n",
    "    elif altLenAdj>0:\n",
    "        seq=seq[p-7:p+8+altLenAdj]    \n",
    "        seqRef=seq[:7]+ref+seq[7+1:-altLenAdj]\n",
    "        seqAlt=seq[:7]+alt+seq[7+1:]\n",
    "    else:\n",
    "        raise ValueError('ref/alt not allowed yet')\n",
    "    \n",
    "    # if len(seqRef)!=len(seqAlt):\n",
    "        # raise ValueError('ref/alt seqs not dovetailed')\n",
    "        \n",
    "    for i,end in enumerate(range(7,len(seq)-altLenAdj)):\n",
    "        start=end-7\n",
    "        last=end+1\n",
    "        kmer8Ref=seqRef[start:last]\n",
    "        kmer8Alt=seqAlt[start:last]\n",
    "            \n",
    "        yield kmer8Ref,kmer8Alt\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "426c577f-d328-42eb-8537-d67a79999a18",
   "metadata": {},
   "outputs": [],
   "source": [
    "def is_tf_8mer(kmer8):\n",
    "    if kmer8 in tfAllPossible8mers:            return True\n",
    "    else:                                      return False\n",
    "\n",
    "# Set up logic for creating/deleting ets sites\n",
    "BeforeAfter2tfLogic={\n",
    "    (0,0):'Tf Never',\n",
    "    (1,1):'Tf Always',\n",
    "    (0,1):'Tf Created',\n",
    "    (1,0):'Tf Deleted'\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6a028beb-3e66-4c5f-a718-c2f1e9690509",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def snv_tf_analysis(chrom,idx1,ref,alt,chr2seq,tf2aff):\n",
    "    \n",
    "    # Ensure data is snp, no indels\n",
    "    prohibitedPolyTypes=['Other']\n",
    "    if len(ref)==1 and len(alt)==1: \n",
    "        polymorphismType='SNV'\n",
    "        \n",
    "    elif len(ref)==1 and len(alt)==0:\n",
    "        polymorphismType='SND'\n",
    "        \n",
    "    else:\n",
    "        polymorphismType='Other'\n",
    "        \n",
    "    if polymorphismType in prohibitedPolyTypes:\n",
    "        return 'Variant Type Not Allowed Yet'\n",
    "        \n",
    "    ######################################\n",
    "    # Verify ref/alt matches genomic bp\n",
    "    ######################################\n",
    "    \n",
    "    if 'chr' not in chrom: chrom=f'chr{chrom}'\n",
    "    \n",
    "    # Check if pos in genome is same as declared REF\n",
    "    idx0=int(idx1)-1\n",
    "    snvNtInGenome=chr2seq[chrom][idx0]\n",
    "    if ref==snvNtInGenome:    \n",
    "        checkSnv='REF'\n",
    "    elif alt==snvNtInGenome:  \n",
    "        checkSnv='ALT'\n",
    "    else:                     \n",
    "        return f'Genomic bp does not match expectation: \\n  expected ref/alt = ({ref}/{alt})\\n  observed = {snvNtInGenome}'\n",
    "    \n",
    "    \n",
    "    ######################################\n",
    "    # Assess if SNV is in ETS\n",
    "    ######################################\n",
    "    \n",
    "    tfChanged=False\n",
    "    for kmer8,kmer8mut in getAll8mers(chr2seq[chrom],idx0,ref,alt):\n",
    "\n",
    "        ba=(is_tf_8mer(kmer8),is_tf_8mer(kmer8mut)) # before after logic\n",
    "        tfLogic=BeforeAfter2tfLogic[ba]\n",
    "\n",
    "        # If ets-changing snp found, write it out\n",
    "        if tfLogic=='Tf Always': \n",
    "            tfChanged=True\n",
    "            refaff=tf2aff[kmer8]\n",
    "            altaff=tf2aff[kmer8mut]\n",
    "            \n",
    "            return tfLogic,refaff,altaff,altaff/refaff,kmer8,kmer8mut\n",
    "        \n",
    "        elif tfLogic=='Tf Created':\n",
    "            return tfLogic,kmer8,kmer8mut\n",
    "            \n",
    "        elif tfLogic=='Tf Deleted':\n",
    "            return tfLogic,kmer8,kmer8mut\n",
    "\n",
    "    if tfChanged==False:\n",
    "        return 'Variant not in Tf',kmer8,kmer8mut\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6e19887a-63bf-40ba-b1c1-8129df745feb",
   "metadata": {},
   "source": [
    "# Load ref data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2ea6fbb1-3d7b-4be4-b725-9c0ac193f93c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ap1\n",
    "ref  =f'preprocessed-data//parsed_Jun_Fos_8mers_core=TGANTMA_max=TGAGTCAT.txt'\n",
    "Seq2EtsAff  = hf.loadAff(ref)\n",
    "hf.dprint(Seq2EtsAff)\n",
    "\n",
    "\n",
    "# hg38\n",
    "ref=f'/Users/joe/code/ref/genomes/human/hg38/fasta/hg38.fa'\n",
    "chr2seq=hf.faLoadGenome(ref)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9975c417-0e0f-4efa-a339-679a983a6bce",
   "metadata": {},
   "source": [
    "# Analysis options"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "87060946-53ab-4f71-acb9-d4db13e2b820",
   "metadata": {},
   "outputs": [],
   "source": [
    "fcOptCutoff=1.5\n",
    "# fcOptCutoff=2\n",
    "# fcOptCutoff=3.46\n",
    "# fcOptCutoff=5"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0d8b52f4-6577-4367-9934-e58188a9932c",
   "metadata": {},
   "source": [
    "# Load MPRA data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4082b2ab-fe19-4e47-bf26-e302880d3a37",
   "metadata": {},
   "outputs": [],
   "source": [
    "fn=f'preprocessed-data/QTD000110__readyForInput__matteoCorrMethod_keepOnlyTewheyMPRAVariants=False_removeMixedGofLofVariants=False_genomeWideSigAdj=True.tsv'\n",
    "\n",
    "df=pd.read_csv(fn,sep='\\t')\n",
    "df.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "99e06545-5260-42af-a770-db82ccfe764e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# only snvs\n",
    "df=df.loc[(df.ref.str.len()==1) & (df.alt.str.len()==1),:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "01eb846c-9e60-4415-a29d-b4a39940cf85",
   "metadata": {},
   "outputs": [],
   "source": [
    "len(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c3ce6a89-cf82-462b-8eb6-48c9688fa3c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "## New one which grabs 8mer ref and  8mer alt  regardlesso f snv  type\n",
    "\n",
    "variantTypeList=[]\n",
    "refAffList=[]\n",
    "altAffList=[]\n",
    "fcList=[]\n",
    "ref8merlist=[]\n",
    "alt8merlist=[]\n",
    "\n",
    "for idx in df.index:\n",
    "    \n",
    "    chrom,pos,ref,alt=[df.at[idx,ci] for ci in ['chrom','pos','ref','alt']]\n",
    "\n",
    "    if alt=='-': alt=''\n",
    "    resultList=snv_tf_analysis(str(chrom),pos,ref,alt,chr2seq,Seq2EtsAff)\n",
    "    result=resultList[0]\n",
    "    \n",
    "    if result=='Variant not in Tf':\n",
    "        result,kmer8,kmer8mut=resultList\n",
    "        variantTypeList.append('Not In Tf')\n",
    "        refAffList.append(np.NaN)\n",
    "        altAffList.append(np.NaN)\n",
    "        fcList.append(np.NaN)\n",
    "        ref8merlist.append(kmer8)\n",
    "        alt8merlist.append(kmer8mut)\n",
    "\n",
    "\n",
    "    elif result=='Tf Created':\n",
    "        result,kmer8,kmer8mut=resultList\n",
    "        variantTypeList.append('Tf Created')\n",
    "        refAffList.append(np.NaN)\n",
    "        altAffList.append(np.NaN)\n",
    "        fcList.append(np.NaN)\n",
    "        ref8merlist.append(kmer8)\n",
    "        alt8merlist.append(kmer8mut)\n",
    "\n",
    "\n",
    "    elif result=='Tf Deleted':\n",
    "        result,kmer8,kmer8mut=resultList\n",
    "        variantTypeList.append('Tf Deleted')\n",
    "        refAffList.append(np.NaN)\n",
    "        altAffList.append(np.NaN)\n",
    "        fcList.append(np.NaN)\n",
    "        ref8merlist.append(kmer8)\n",
    "        alt8merlist.append(kmer8mut)\n",
    "\n",
    "        \n",
    "    else:\n",
    "        variantType,refAff,altAff,fc,kmer8ref,kmer8alt=resultList\n",
    "        variantTypeList.append(variantType)\n",
    "        refAffList.append(refAff)\n",
    "        altAffList.append(altAff)\n",
    "        fcList.append(fc)\n",
    "        ref8merlist.append(kmer8ref)\n",
    "        alt8merlist.append(kmer8alt)\n",
    "\n",
    "        \n",
    "        \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "98d96818-7200-4207-a262-2b99f7071f04",
   "metadata": {},
   "outputs": [],
   "source": [
    "df['variant-type']=variantTypeList\n",
    "df['aff-ref']=refAffList\n",
    "df['aff-alt']=altAffList\n",
    "df['aff-fc']=fcList\n",
    "df['kmer-ref']=ref8merlist\n",
    "df['kmer-alt']=alt8merlist\n",
    "\n",
    "df['aff-delta']=df.apply(lambda row: row['aff-alt']-row['aff-ref'],axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "364f8011-616b-463a-892a-84aa3d903029",
   "metadata": {},
   "outputs": [],
   "source": [
    "df['variant-type'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "944e4c5c-a546-475d-9796-85a71656b61b",
   "metadata": {},
   "outputs": [],
   "source": [
    "df=df.loc[df['pvalue-bonf']<.01,:]\n",
    "df['-log(p)']=df['pvalue-bonf'].apply(lambda p: -np.log10(p))\n",
    "df['-log(p)*direction']=df.apply(lambda r: r['-log(p)'] if r['beta']>=0 else -r['-log(p)'],axis=1)\n",
    "df['-log(p)*direction']=df['-log(p)*direction'].replace(-0,0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ee81efab-4447-48f9-81c4-5992709e9e0f",
   "metadata": {},
   "outputs": [],
   "source": [
    "df['pvalue-bonf'].max()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "352701e4-17fe-4036-92b7-c5eee656cedd",
   "metadata": {},
   "outputs": [],
   "source": [
    "df['pvalue-bonf'].min()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6422b263-8ca8-4f28-b5c0-b05c37ac9681",
   "metadata": {},
   "outputs": [],
   "source": [
    "df['variant-type'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "90505d7a-cced-4301-82d1-dec97f496938",
   "metadata": {},
   "outputs": [],
   "source": [
    "df['chrompos']=df['chrom'].apply(lambda c: 'chr'+str(c))+df['pos'].apply(str)\n",
    "print(len(df['chrompos']))\n",
    "print(len(df['chrompos'].unique()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e2f40ba7-a7df-44f7-b199-bbd482c9d51c",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(len(df))\n",
    "df=df.drop_duplicates()\n",
    "print(len(df))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5885d241-7040-4eb6-911e-3296eef29a26",
   "metadata": {},
   "source": [
    "# Prediction accuracy vs random (for nature)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4fda9015-c8aa-4be2-909d-9d2155455490",
   "metadata": {},
   "outputs": [],
   "source": [
    "pvalCutoff=.05\n",
    "logSigCutoff=abs(np.log10(pvalCutoff))\n",
    "\n",
    "deNovoThresh=.25\n",
    "\n",
    "fcOptCutoff=1.5\n",
    "\n",
    "fcDecCutoff=sf.round(1/fcOptCutoff,3)\n",
    "\n",
    "fcNoChangeCutoff_Inc=1.25\n",
    "fcNoChangeCutoff_Dec=sf.round(1/fcNoChangeCutoff_Inc,3)\n",
    "\n",
    "# get variants to predict on\n",
    "predDF=df.loc[:,:]\n",
    "\n",
    "# assign prediction labels\n",
    "                           # make this ever so slightly lower than 0 to avoid having bin edges be the same\n",
    "bins=[-np.inf,-logSigCutoff-1e-100,logSigCutoff,np.inf]\n",
    "predDF['classification']=pd.cut(predDF['-log(p)*direction'],bins=bins,labels=['Sig Decrease','N.S.','Sig Increase'])\n",
    "\n",
    "# assign predictions\n",
    "etsLabels=[]\n",
    "\n",
    "for fc,classification,kmer,kmerMut in hf.zipdf(predDF,['aff-fc','variant-type','kmer-ref','kmer-alt']):\n",
    "\n",
    "    # not in tf\n",
    "    if classification=='Not In Tf':\n",
    "        anno='not-in-tf'\n",
    "\n",
    "    # delet tf\n",
    "    elif classification=='Tf Deleted':\n",
    "        anno='del-tf'\n",
    "\n",
    "    elif classification=='Tf Created':\n",
    "\n",
    "        denovoaff=Seq2EtsAff[kmerMut]\n",
    "\n",
    "        if denovoaff<deNovoThresh:\n",
    "            anno=f'de-novo-tf-low-aff\\n<{deNovoThresh}'            \n",
    "\n",
    "        elif denovoaff>=deNovoThresh:\n",
    "            anno=f'de-novo-tf-high-aff\\n>={deNovoThresh}'\n",
    "\n",
    "    #  dramatic iincrase\n",
    "    elif fc>=fcOptCutoff:\n",
    "        anno=f'dramatic-increase\\n>={fcOptCutoff}'\n",
    "\n",
    "    #  dramatic incrase\n",
    "    elif fc<=fcDecCutoff:\n",
    "        # print(fc,classification)\n",
    "        anno=f'dramatic-decrease\\n<={fcDecCutoff}'\n",
    "\n",
    "    # no increase\n",
    "    elif fc<=fcNoChangeCutoff_Inc and fc>=fcNoChangeCutoff_Dec:\n",
    "        anno=f'no-change-aff\\n<={fcNoChangeCutoff_Inc}\\n>={fcNoChangeCutoff_Dec}'\n",
    "\n",
    "    else:\n",
    "        # print(fc,classification,kmer,kmerMut)\n",
    "        anno='subtle-change'\n",
    "\n",
    "\n",
    "    etsLabels.append(anno)\n",
    "\n",
    "predDF['tf-class']=etsLabels\n",
    "\n",
    "line_out=hf.write_row(['snv-group','total','effect-ns','effect-gof','effect-lof'])\n",
    "\n",
    "for classi in predDF['tf-class'].unique():\n",
    "    nClassi=predDF['tf-class'].value_counts()[classi]\n",
    "\n",
    "    ns,gof,lof=[predDF.loc[predDF['tf-class']==classi,'classification'].value_counts()[ci] for ci in ['N.S.','Sig Increase','Sig Decrease']]\n",
    "\n",
    "    line_out+=hf.write_row([classi.replace('\\n',' '),nClassi,ns,gof,lof])\n",
    "\n",
    "print(line_out)\n",
    "with open(f'FOR-NATURE-STATS-PREDICTIONS-{expname}-{tfName}__fdr.tsv','w') as f: f.write(line_out)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "51342b9c-7837-4650-9203-6402c397b40a",
   "metadata": {},
   "outputs": [],
   "source": [
    "altHypothesis='less'\n",
    "\n",
    "datalabels_subset_v2=[\n",
    "[f'not-in-tf',f'del-tf',f'dramatic-decrease\\n<={fcDecCutoff}',f'de-novo-tf-low-aff\\n<{deNovoThresh}',f'de-novo-tf-high-aff\\n>={deNovoThresh}','subtle-change'],\n",
    "[f'no-change-aff\\n<={fcNoChangeCutoff_Inc}\\n>={fcNoChangeCutoff_Dec}'],\n",
    "[f'dramatic-increase\\n>={fcOptCutoff}'],\n",
    "]\n",
    "\n",
    "data2color={\n",
    "(f'no-change-aff\\n<={fcNoChangeCutoff_Inc}\\n>={fcNoChangeCutoff_Dec}',):'lightgrey',\n",
    "(f'dramatic-increase\\n>={fcOptCutoff}',):'forestgreen',\n",
    "(f'not-in-tf',f'del-tf',f'dramatic-decrease\\n<={fcDecCutoff}',f'de-novo-tf-low-aff\\n<{deNovoThresh}',f'de-novo-tf-high-aff\\n>={deNovoThresh}','subtle-change'):'dimgrey'}\n",
    "\n",
    "\n",
    "plotDF=predDF.loc[:,:].reset_index().copy(deep=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9dc06a05-dc06-4d08-b2c5-b336613ede55",
   "metadata": {},
   "outputs": [],
   "source": [
    "rsid2mostsigrow={}\n",
    "\n",
    "for i,(rsid,pvalue) in enumerate(hf.zipdf(plotDF,['rsid','pvalue-bonf'])):\n",
    "    \n",
    "    if rsid not in rsid2mostsigrow:\n",
    "        rsid2mostsigrow[rsid]=i\n",
    "        \n",
    "    else:\n",
    "        #update if this is more sig\n",
    "        if pvalue<rsid2mostsigrow[rsid]:\n",
    "            rsid2mostsigrow[rsid]=i\n",
    "\n",
    "idxToKeep=rsid2mostsigrow.values()\n",
    "plotDF=plotDF.loc[idxToKeep,:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a6336426-a0ee-48dd-aed1-3dab339b5d5d",
   "metadata": {},
   "outputs": [],
   "source": [
    "violin=True\n",
    "boxplot=False\n",
    "\n",
    "data=[]\n",
    "for di in datalabels_subset_v2:\n",
    "    datai=[]\n",
    "    for dj in di:\n",
    "        datai+=plotDF.loc[(plotDF['tf-class']==dj),'-log(p)*direction'].tolist()\n",
    "    data.append(datai)\n",
    "\n",
    "line_out=hf.write_row(['group1','group2','p'])\n",
    "\n",
    "############################################################################\n",
    "# Stats\n",
    "############################################################################\n",
    "\n",
    "dCtrlNameList=datalabels_subset_v2[0]\n",
    "dCtrl=plotDF.loc[(plotDF['tf-class'].isin(dCtrlNameList)),'-log(p)*direction']\n",
    "\n",
    "for diNameList in datalabels_subset_v2[1:]:\n",
    "\n",
    "    di=plotDF.loc[(plotDF['tf-class'].isin(diNameList)),'-log(p)*direction']\n",
    "\n",
    "    try:\n",
    "        _ , p =stats.mannwhitneyu(dCtrl,di,alternative=altHypothesis)\n",
    "    except ValueError:\n",
    "        pass\n",
    "\n",
    "\n",
    "    line_out+=hf.write_row(['+'.join(dCtrlNameList).replace('\\n',' '),'+'.join(diNameList).replace('\\n',' '),sf.round(p,4)])\n",
    "\n",
    "print(line_out)\n",
    "with open(f'FOR-NATURE__violin-stats__{tfName}__{expname}.tsv','w') as f: f.write(line_out)\n",
    "\n",
    "############################################################################\n",
    "# Stats\n",
    "############################################################################\n",
    "\n",
    "fig,ax=plt.subplots(1,1,figsize=(2*len(datalabels_subset_v2),4),dpi=300,facecolor='white')\n",
    "\n",
    "palette=[data2color[tuple(di)] for di in datalabels_subset_v2]\n",
    "\n",
    "if violin:\n",
    "    sns.violinplot(data=data,color='white',inner=None,cut=0,ax=ax,width=.5,scale='width')\n",
    "\n",
    "    meanlinewidth=.3\n",
    "    for xi,di in enumerate(data):\n",
    "        val=np.mean(di)\n",
    "        ax.plot([xi-meanlinewidth,xi+meanlinewidth],[val,val],color='grey')\n",
    "\n",
    "if boxplot:\n",
    "    sns.boxplot(data=data,color='white',ax=ax,fliersize=0)#,palette=palette)\n",
    "\n",
    "ax.spines['top'].set_visible(False)\n",
    "ax.spines['right'].set_visible(False)\n",
    "\n",
    "axticklabels=['\\n'.join(label) for label in datalabels_subset_v2]\n",
    "ax.set_xticklabels([f'{label}\\nN={len(points):,}' for label,points in zip(axticklabels,data)])\n",
    "ax.set_ylabel('-log(bh_fdr)*direction')\n",
    "\n",
    "ax.axhline(-np.log10(.05),ls='--',color='green',alpha=.2)\n",
    "ax.axhline(np.log10(.05),ls='--',color='red',alpha=.2)\n",
    "ax.axhline(np.log10(1),ls='--',color='black',alpha=.2)\n",
    "\n",
    "ax.set_title(f'{expname} fcOptCutoff={fcOptCutoff}')\n",
    "\n",
    "ax.set_ylim(-45,45)\n",
    "\n",
    "\n",
    "# plt.savefig(f'/Users/joe/Desktop/ets-kircher-elements={idList}.png')# ,bbox_inches='tight')\n",
    "plt.show()\n",
    "\n",
    "plotDF.to_csv(f'NATURE-ALL-DATA_{tfName}_{expname}.tsv',sep='\\t',index=None)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
