{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f9b92555-6a5a-4cb1-a8de-554b20d58d13",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#!/usr/bin/env python\n",
    "# coding: utf-8\n",
    "\n",
    "print('reading libraries...')\n",
    "\n",
    "from os.path import exists\n",
    "from bx.intervals.io import GenomicIntervalReader\n",
    "from bx.bbi.bigwig_file import BigWigFile\n",
    "import numpy as np\n",
    "import time\n",
    "import sys\n",
    "import seaborn as sns\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import scipy.stats as stats\n",
    "import pickle \n",
    "import helper_functions as hf\n",
    "import sigfig as sf\n",
    "\n",
    "def get_list_from_counts(value2cont):\n",
    "    out=[]\n",
    "    for v,c in value2cont.items():\n",
    "        out+=[v]*c\n",
    "    return out\n",
    "\n",
    "# Inputs\n",
    "inDir='data-outputted/'\n",
    "\n",
    "fn2genome={\n",
    "    'GSM2218592_B_Ets1_R1_bow':'mm9', # b cells, Saelee P et al\n",
    "    'GSM4110116_Ets1_S0_1_129929':'hg19', # t cells, McCarter AC et al\n",
    "    'GSM3520734_NK_ETS1_chip':'hg38', # nk cells, Taveirne S et al\n",
    "}\n",
    "\n",
    "print('loading genomes...')\n",
    "genome2chr2seq={\n",
    "    'hg38':hf.faLoadGenome('hg38.fa'),\n",
    "    'hg19':hf.faLoadGenome('hg19.fa'),\n",
    "    'mm9' :hf.faLoadGenome('mm9.fa')\n",
    "}\n",
    "\n",
    "# exit early to plot fast\n",
    "beta=False\n",
    "\n",
    "# set up plot parameters\n",
    "setYlimCustom=False\n",
    "ymax=3\n",
    "ymin=0\n",
    "\n",
    "# remove mitochondrial chromosome\n",
    "genome2chr2seq[genome].pop('chrM')\n",
    "\n",
    "for fn,genome in fn2genome.items():\n",
    "    \n",
    "    print(f'loading {fn} bigwig...')\n",
    "    bw = BigWigFile( open( f'../../data-downloaded/{fn}.bigWig' ,'rb') )\n",
    "    \n",
    "    # initialize dataset\n",
    "    merged_Chr2Kmer2Data={}\n",
    "\n",
    "    print('\\tloading ets kmer data...')\n",
    "    \n",
    "    # remove non chromosomal contigs from genome \n",
    "    rm_contigs = []\n",
    "    for i in genome2chr2seq[genome].keys():\n",
    "        if i[0:3] != 'chr':\n",
    "            rm_contigs.append(i)\n",
    "    if len(rm_contigs) != 0:\n",
    "        for x in rm_contigs:\n",
    "            genome2chr2seq[genome].pop(x)\n",
    "    \n",
    "    \n",
    "    for i,chrom in enumerate(genome2chr2seq[genome]):\n",
    "\n",
    "        merged_Chr2Kmer2Data[chrom]={}\n",
    "        \n",
    "        # load data processed in FS2ABi-Preprocess-Data.py\n",
    "        fnPickle=f'{inDir}/{fn}/{fn}__Kmer2Data__beta=False__chrom={chrom}.pydict.pickle'\n",
    "        \n",
    "        if exists(fnPickle)==False: continue\n",
    "        \n",
    "        with open(fnPickle,'rb') as f: \n",
    "            Kmer2Data=pickle.load(f)\n",
    "        \n",
    "        # iterate over data\n",
    "        for kmer in Kmer2Data:\n",
    "\n",
    "            if kmer not in merged_Chr2Kmer2Data[chrom]:\n",
    "                merged_Chr2Kmer2Data[chrom][kmer]={'pbm-aff':Kmer2Data[kmer]['pbm-aff']}\n",
    "                merged_Chr2Kmer2Data[chrom][kmer]['ChipSignal2Count']={}\n",
    "\n",
    "            for chipSignal in Kmer2Data[kmer]['chip-signal-list']:\n",
    "\n",
    "                if chipSignal not in merged_Chr2Kmer2Data[chrom][kmer]['ChipSignal2Count']:\n",
    "                    merged_Chr2Kmer2Data[chrom][kmer]['ChipSignal2Count'][chipSignal]=0\n",
    "                merged_Chr2Kmer2Data[chrom][kmer]['ChipSignal2Count'][chipSignal]+=1\n",
    "\n",
    "        if i>1 and beta: break\n",
    "        \n",
    "    # bin ets sites by their affinity (ie .1, .2, ... 1.0)\n",
    "    print('\\tbinning data...')\n",
    "    bins=np.arange(0,1.1,.1)\n",
    "\n",
    "    c2v={'bin':[],'signal':[],'chrom':[]}\n",
    "\n",
    "    for i,chrom in enumerate(merged_Chr2Kmer2Data):\n",
    "\n",
    "        for kmer in merged_Chr2Kmer2Data[chrom]:\n",
    "            binIdx=np.digitize(merged_Chr2Kmer2Data[chrom][kmer]['pbm-aff'], bins,right=True)\n",
    "            merged_Chr2Kmer2Data[chrom][kmer]['bin']=bins[binIdx]\n",
    "\n",
    "        x2y={} # mapping xposition to y value of binned data at that xposition\n",
    "        \n",
    "        for kmer in merged_Chr2Kmer2Data[chrom]:\n",
    "            x=merged_Chr2Kmer2Data[chrom][kmer]['bin']\n",
    "            y=get_list_from_counts(merged_Chr2Kmer2Data[chrom][kmer]['ChipSignal2Count'])\n",
    "\n",
    "            if x not in x2y:\n",
    "                x2y[x]=[]\n",
    "\n",
    "            x2y[x]+=y\n",
    "\n",
    "\n",
    "        for xi,yiList in x2y.items():\n",
    "            for yi in yiList:\n",
    "                c2v['bin'].append(round(xi,3))\n",
    "                c2v['signal'].append(yi)\n",
    "                c2v['chrom'].append(chrom)\n",
    "\n",
    "\n",
    "        if i>1 and beta: break\n",
    "\n",
    "    # convert to dataframe\n",
    "    df=pd.DataFrame(c2v)\n",
    "    \n",
    "    # drop infinites and na\n",
    "    df=df.replace(np.inf, np.nan).replace(-np.inf, np.nan).dropna()\n",
    "\n",
    "    # begin plot\n",
    "    fig,ax=plt.subplots(1,2,figsize=(13,3),dpi=300,gridspec_kw={'width_ratios': [5,1]})\n",
    "\n",
    "\n",
    "    #####################################################################\n",
    "    # averages\n",
    "    #####################################################################\n",
    "\n",
    "    avgFunc=np.mean\n",
    "    errFunc=stats.sem\n",
    "\n",
    "    minAvg=+np.inf\n",
    "    maxAvg=-np.inf\n",
    "\n",
    "\n",
    "    avgList=[]\n",
    "    errList=[]\n",
    "    biList=[]\n",
    "    space=.03\n",
    "\n",
    "    # plot average of each column\n",
    "    swarmDF={c:[] for c in ['bin','agg-signal','chrom']}\n",
    "    for chrom in df.chrom.unique():\n",
    "        for bi in sorted(df.bin.unique()):\n",
    "\n",
    "            biData=df.loc[(df['bin']==bi) & (df['chrom']==chrom),'signal']\n",
    "            biAvg=avgFunc(biData)\n",
    "            biErr=errFunc(biData)\n",
    "\n",
    "            avgList.append(biAvg)\n",
    "            errList.append(biErr)\n",
    "\n",
    "            if biAvg<minAvg: minAvg=biAvg\n",
    "            if biAvg>maxAvg: maxAvg=biAvg\n",
    "\n",
    "            biList .append(float(bi))\n",
    "\n",
    "            swarmDF['bin'].append(bi)\n",
    "            swarmDF['agg-signal'].append(biAvg)\n",
    "            swarmDF['chrom'].append(chrom)\n",
    "\n",
    "    swarmDF=pd.DataFrame(swarmDF)\n",
    "\n",
    "    # set custon y-axis limits \n",
    "    if setYlimCustom:\n",
    "        ax[0].set_ylim(ymin,ymax)\n",
    "\n",
    "    sns.swarmplot(x='bin',y='agg-signal',data=swarmDF,ax=ax[0],palette='Blues',size=2)\n",
    "\n",
    "\n",
    "        \n",
    "    #####################################################################\n",
    "    # trendlines\n",
    "    #####################################################################\n",
    "\n",
    "    pround=10\n",
    "\n",
    "    sCorrList=[]\n",
    "    pCorrList=[]\n",
    "\n",
    "    for endPoint in bins[2:]:\n",
    "\n",
    "        corrDF=swarmDF.loc[swarmDF['bin']<=endPoint ,:]\n",
    "        corrDF=corrDF[~corrDF.isin([np.nan, np.inf, -np.inf]).any(1)]\n",
    "        corrDF=corrDF.replace(np.inf, np.nan).replace(-np.inf, np.nan).dropna()\n",
    "\n",
    "        pr,pp=[round(i,pround) for i in stats.pearsonr( corrDF['bin'], corrDF['agg-signal']) ]\n",
    "        sr,sp=[round(i,pround) for i in stats.spearmanr(corrDF['bin'], corrDF['agg-signal']) ]\n",
    "\n",
    "        sCorrList.append((sr,sp))\n",
    "        pCorrList.append((pr,pp))\n",
    "\n",
    "    fitDF=swarmDF.loc[~swarmDF['chrom'].isin(['chrM','chrY']),:].dropna(how='any')\n",
    "\n",
    "    a, b = np.polyfit(fitDF['bin'],fitDF['agg-signal'], 1)\n",
    "\n",
    "    X=range(len(swarmDF['bin'].unique()))\n",
    "    Y= a * np.array(sorted(swarmDF['bin'].unique())) + b\n",
    "\n",
    "\n",
    "    ax[0].plot(X,Y,zorder=100,lw=1,color='blue')      \n",
    "                \n",
    "\n",
    "    #####################################################################\n",
    "    # labels\n",
    "    #####################################################################\n",
    "\n",
    "\n",
    "    ax[0].spines['top'].set_visible(None)\n",
    "    ax[0].spines['right'].set_visible(None)\n",
    "    ax[0].set_xlabel('Affinity Bin')\n",
    "    ax[0].set_ylabel('ETS ChIP-seq Signal\\nOver GGAW 8mer')\n",
    "\n",
    "    # change scale\n",
    "    if not setYlimCustom:\n",
    "        ymin=minAvg-.5\n",
    "        ymax=maxAvg+.5\n",
    "    ax[0].set_ylim(ymin,ymax)\n",
    "    \n",
    "    xticklabels=[]\n",
    "    sigfigRhoRound=3\n",
    "    for (pr,pp),(sr,sp),xlabi in zip([('','')]+pCorrList,[('','')]+sCorrList,bins[1:]):\n",
    "        if pr!='':\n",
    "            newLab=f'{sf.round(xlabi,1)}\\np={sf.round(pr,sigfigRhoRound)}\\ns={sf.round(sr,sigfigRhoRound)}'\n",
    "        else:\n",
    "            newLab=f'{sf.round(xlabi,1)}'\n",
    "        xticklabels.append(newLab)\n",
    "        \n",
    "    ax[0].set_xticklabels(xticklabels)\n",
    "\n",
    "    \n",
    "    #####################################################################\n",
    "    # Noises from random kmers\n",
    "    #####################################################################\n",
    "\n",
    "    # # Add \"noise level\" from random kmers\n",
    "    print('\\tgenerating random kmer data...')\n",
    "    genomeLen,Chrom2NumStartEnd=hf.generate_Chrom2NumStartEnd(genome2chr2seq[genome])\n",
    "\n",
    "    chipSignalRandomKmers=[]\n",
    "    \n",
    "    if beta:randomSamplesize=1000\n",
    "    else:   randomSamplesize=int(len(df)/1000) # there are 10 bins, so take average  bin size\n",
    "\n",
    "    allowedChroms=set(swarmDF['chrom'])\n",
    "    randSwarmDF={c:[]  for c in  ['chrom','signal']}\n",
    "    for i in range(randomSamplesize):\n",
    "\n",
    "        loop=True\n",
    "        while loop:\n",
    "            chrom,start=hf.choose_random_chrom_pos(genomeLen,Chrom2NumStartEnd)\n",
    "            \n",
    "            if chrom not in allowedChroms: continue\n",
    "            \n",
    "            end=start+8\n",
    "            seq=genome2chr2seq[genome][chrom][start:end]\n",
    "\n",
    "            # require real seq\n",
    "            if 'N' in seq: continue\n",
    "\n",
    "            # requre non-ets site\n",
    "            for etsCore in ['GGAA','GGAT','TTCC','ATCC']:\n",
    "                if etsCore in seq: continue\n",
    "\n",
    "            # if all is well, parse chip signal\n",
    "            result=bw.query(chrom, start, start+8, 1)\n",
    "            if result!=None:\n",
    "                result=result[0]['mean']\n",
    "\n",
    "                if str(result)!='nan':\n",
    "                    chipSignalRandomKmers.append(result)\n",
    "                    randSwarmDF['chrom'].append(chrom)\n",
    "                    randSwarmDF['signal'].append(result)\n",
    "                    loop=False\n",
    "\n",
    "    randSwarmDF=pd.DataFrame(randSwarmDF)\n",
    "\n",
    "    randSwarmDF=randSwarmDF.groupby('chrom').apply(avgFunc).reset_index()\n",
    "\n",
    "    ax[1].set_xlim(-1,1)\n",
    "\n",
    "    sns.swarmplot(y='signal',data=randSwarmDF,ax=ax[1],size=3,color='lightgrey')\n",
    "\n",
    "    ax[1].spines['top'].set_visible(None)\n",
    "    ax[1].spines['right'].set_visible(None)\n",
    "    ax[1].set_xlabel(f'Random Non-ETS Kmers\\nN={randomSamplesize:,}')\n",
    "    ax[1].set_ylabel('ETS ChIP-seq Signal')\n",
    "\n",
    "    ax[1].set_ylim(ymin,ymax)\n",
    "\n",
    "    ax[1].set_xticks([0])\n",
    "    \n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b313019d-6d03-49b0-98df-9d7153855feb",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
