{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "2d7471c1-c049-4157-9e6b-95c22fc5cc4f",
   "metadata": {},
   "source": [
    "# Method to analyze whether SNV in ETS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0f155b95-9f5f-43e9-ae06-94151b69fb01",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "import helper_functions as hf\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import pickle\n",
    "import os\n",
    "import scipy.stats as stats\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import sigfig as sf\n",
    "\n",
    "import statsmodels.api as sm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "03b20a1a-d925-45f6-a023-b6d15c126422",
   "metadata": {},
   "outputs": [],
   "source": [
    "def getAll8mers(seq,p,ref,alt):\n",
    "    \n",
    "    altLenAdj=len(ref)-len(alt)\n",
    "        \n",
    "    if altLenAdj==0:\n",
    "        seq=seq[p-7:p+8]    \n",
    "        seqRef=seq[:7]+ref+seq[7+1:]\n",
    "        seqAlt=seq[:7]+alt+seq[7+1:]\n",
    "        \n",
    "    elif altLenAdj>0:\n",
    "        seq=seq[p-7:p+8+altLenAdj]    \n",
    "        seqRef=seq[:7]+ref+seq[7+1:-altLenAdj]\n",
    "        seqAlt=seq[:7]+alt+seq[7+1:]\n",
    "    else:\n",
    "        raise ValueError('ref/alt not allowed yet')\n",
    "    \n",
    "    # if len(seqRef)!=len(seqAlt):\n",
    "        # raise ValueError('ref/alt seqs not dovetailed')\n",
    "        \n",
    "    for i,end in enumerate(range(7,len(seq)-altLenAdj)):\n",
    "        start=end-7\n",
    "        last=end+1\n",
    "        kmer8Ref=seqRef[start:last]\n",
    "        kmer8Alt=seqAlt[start:last]\n",
    "            \n",
    "        yield kmer8Ref,kmer8Alt\n",
    "            \n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "426c577f-d328-42eb-8537-d67a79999a18",
   "metadata": {},
   "outputs": [],
   "source": [
    "def ets(kmer4):\n",
    "    if kmer4 in ['GGAA','GGAT','TTCC','ATCC']: return True\n",
    "    else:                                      return False\n",
    "\n",
    "# Set up logic for creating/deleting ets sites\n",
    "BeforeAfter2EtsLogic={\n",
    "    (0,0):'Ets Never',\n",
    "    (1,1):'Ets Always',\n",
    "    (0,1):'Ets Created',\n",
    "    (1,0):'Ets Deleted'\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4321a122-032e-4d6f-80d8-55ff0c81c292",
   "metadata": {},
   "outputs": [],
   "source": [
    "def snv_ets_analysis(chrom,idx1,ref,alt,chr2seq,tf2aff):\n",
    "    \n",
    "    # Ensure data is snp, no indels\n",
    "    prohibitedPolyTypes=['Other']\n",
    "    if len(ref)==1 and len(alt)==1: \n",
    "        polymorphismType='SNV'\n",
    "        \n",
    "    elif len(ref)==1 and len(alt)==0:\n",
    "        polymorphismType='SND'\n",
    "        \n",
    "    else:\n",
    "        polymorphismType='Other'\n",
    "        \n",
    "    if polymorphismType in prohibitedPolyTypes:\n",
    "        return 'Variant Type Not Allowed Yet'\n",
    "        \n",
    "    ######################################\n",
    "    # Verify ref/alt matches genomic bp\n",
    "    ######################################\n",
    "    \n",
    "    if 'chr' not in chrom: chrom=f'chr{chrom}'\n",
    "    \n",
    "    # Check if pos in genome is same as declared REF\n",
    "    idx0=int(idx1)\n",
    "    snvNtInGenome=chr2seq[chrom][idx0]\n",
    "    if ref==snvNtInGenome:    \n",
    "        checkSnv='REF'\n",
    "    elif alt==snvNtInGenome:  \n",
    "        checkSnv='ALT'\n",
    "    else:                     \n",
    "        return f'Genomic bp does not match expectation: \\n  expected ref/alt = ({ref}/{alt})\\n  observed = {snvNtInGenome}'\n",
    "    \n",
    "    \n",
    "    ######################################\n",
    "    # Assess if SNV is in ETS\n",
    "    ######################################\n",
    "    \n",
    "    etsChanged=False\n",
    "    for kmer8,kmer8mut in getAll8mers(chr2seq[chrom],idx0,ref,alt):\n",
    "        core   =kmer8[2:6]\n",
    "        coremut=kmer8mut[2:6]\n",
    "\n",
    "        ba=(ets(core),ets(coremut)) # before after logic\n",
    "        etsLogic=BeforeAfter2EtsLogic[ba]\n",
    "\n",
    "        # If ets-changing snp found, write it out\n",
    "        if etsLogic=='Ets Always': \n",
    "            etsChanged=True\n",
    "            refaff=tf2aff[kmer8]\n",
    "            altaff=tf2aff[kmer8mut]\n",
    "            \n",
    "            return polymorphismType,refaff,altaff,altaff/refaff,kmer8,kmer8mut\n",
    "\n",
    "    if etsChanged==False:\n",
    "        return 'Variant not in ETS'\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0bfabb0c-f200-47e4-b444-4e7b99f25e88",
   "metadata": {},
   "source": [
    "# Load ref data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2ea6fbb1-3d7b-4be4-b725-9c0ac193f93c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Ets1\n",
    "ref  =f'preprocessed-data/parsed_Ets1_8mers.txt'              \n",
    "Seq2EtsAff  = hf.loadAff(ref)\n",
    "hf.dprint(Seq2EtsAff)\n",
    "\n",
    "\n",
    "# hg38\n",
    "ref='/Users/joe/code/ref/genomes/human/hg38/fasta/hg38.fa'\n",
    "chr2seq=hf.faLoadGenome(ref)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0d8b52f4-6577-4367-9934-e58188a9932c",
   "metadata": {},
   "source": [
    "# Load MPRA data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4082b2ab-fe19-4e47-bf26-e302880d3a37",
   "metadata": {},
   "outputs": [],
   "source": [
    "fn='preprocessed-data/kvon-2020-table-muts-analyzed.tsv'\n",
    "df=pd.read_csv(fn,sep='\\t')\n",
    "\n",
    "df=df.set_index('VISTAID')\n",
    "df['SEQUENCE']=df['SEQUENCE'].apply(lambda s: s.upper())\n",
    "\n",
    "zrs_seq=df.at['Hs2496','SEQUENCE']\n",
    "\n",
    "df.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f1c5bb9f-f64c-4224-8182-e775f626331d",
   "metadata": {},
   "outputs": [],
   "source": [
    "zrs_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "93bc7282-40e1-4783-9f15-4cc174b4c496",
   "metadata": {},
   "outputs": [],
   "source": [
    "# method to get ets starts in a sequence\n",
    "def get_ets_starts(seq):\n",
    "    starts=[]\n",
    "    for i,kmer in enumerate(hf.get_kmers(seq,8)):\n",
    "        \n",
    "        # if core is ets\n",
    "        if ets(kmer[2:6]):\n",
    "            starts.append(i)\n",
    "            \n",
    "    return set(starts)\n",
    "            \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "163125e9-0de3-45db-a882-f2b12e8e52d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# method to determine which bases arer in an ets\n",
    "def get_ets_bps(seq):\n",
    "    starts=[]\n",
    "    bp2start={}\n",
    "    start2bp={}\n",
    "    for i,kmer in enumerate(hf.get_kmers(seq,8)):\n",
    "        \n",
    "        # if core is ets\n",
    "        if ets(kmer[2:6]):\n",
    "            start2bp[i]=set()\n",
    "            for j in range(i,i+8):\n",
    "                starts.append(j)\n",
    "                start2bp[i].add(j)\n",
    "                bp2start[j]=i\n",
    "            \n",
    "    return set(starts),bp2start,start2bp\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "54a61127-a130-4a0e-89da-ad947cac577a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_mismaches(seq,altseq):\n",
    "    if len(seq)!=len(altseq):\n",
    "        raise ValueError('length mismatch')\n",
    "        \n",
    "    mismatches=[]\n",
    "    for i,(ref,alt) in enumerate(zip(seq,altseq)):\n",
    "        \n",
    "        if ref!=alt:\n",
    "            mismatches.append(i)\n",
    "            \n",
    "    return set(mismatches)\n",
    "         \n",
    "altseq=df.at['Hs2496.75','SEQUENCE']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e75423dc-9952-45d3-8fbc-3dcfaf70412e",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "zrsEtsBases,zrsBp2Start,zrsStart2Bp =get_ets_bps(zrs_seq)\n",
    "zrsEtsStarts=get_ets_starts(zrs_seq)\n",
    "\n",
    "zrsStart2Aff={zrsStart : Seq2EtsAff[ zrs_seq[ zrsStart:zrsStart+8 ] ] for zrsStart in zrsStart2Bp}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "caddfe87-98eb-427f-94b3-c56a67d735a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "Anno2Effect={'normal':0, 'gain':1, 'strong gain':1, 'reduced':-1, 'loss':-1}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "85543f86-2654-4865-ad6f-0058ac3103b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# only consider 2percent muts\n",
    "df=df.loc[df['TYPE']=='2percent']\n",
    "    \n",
    "\n",
    "df['PVALUE LOF']=sm.stats.multipletests(df['PVALUE LOF'],method='fdr_bh')[1]\n",
    "df['PVALUE GOF']=sm.stats.multipletests(df['PVALUE GOF'],method='fdr_bh')[1]\n",
    "    \n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "88e783ff-ee14-488b-ac4a-6c5db629018b",
   "metadata": {},
   "outputs": [],
   "source": [
    "enList=[]\n",
    "pvalueList=[]\n",
    "effectList=[]\n",
    "variantTypeList=[] # Not In Ets, delete,decrease,increase,de novoo\n",
    "refAffList=[]\n",
    "altAffList=[]\n",
    "fcList=[]\n",
    "ref8merlist=[]\n",
    "alt8merlist=[]\n",
    "groupList=[]\n",
    "\n",
    "seqTestedList=[]\n",
    "\n",
    "for idx in df.index:\n",
    "    \n",
    "    group=df.at[idx,'TYPE']\n",
    "    \n",
    "    # only consider the 2percent class\n",
    "    if group not in ['2percent']: continue\n",
    "    \n",
    "    anno=df.at[idx,'ANNOTATION']\n",
    "    seqalt=df.at[idx,'SEQUENCE']\n",
    "    \n",
    "    effect=Anno2Effect[anno]\n",
    "    if   effect>0: pval=df.at[idx,'PVALUE GOF']\n",
    "    elif effect<0: pval=df.at[idx,'PVALUE LOF']\n",
    "    else:          pval=1\n",
    "    \n",
    "    # determine mismatches\n",
    "    mismatchSet=get_mismaches(zrs_seq,seqalt)\n",
    "    \n",
    "    # determine which bases are ets & starts of ets\n",
    "    altEtsBases,altBp2Start,altStart2Bp  = get_ets_bps(seqalt)\n",
    "    altEtsStarts = get_ets_starts(seqalt)\n",
    "    \n",
    "    # determine denovo ets bass\n",
    "    denovoEtsBases = altEtsBases - zrsEtsBases\n",
    "    denovoEtsstarts = altEtsStarts - zrsEtsStarts\n",
    "    \n",
    "    # determine deleted bases\n",
    "    deletedEtsBases = zrsEtsBases - altEtsBases\n",
    "    deletedetsStarts = zrsEtsStarts - altEtsStarts\n",
    "    \n",
    "    # iterate over all ets sites\n",
    "    allEtsStarts = altEtsStarts | zrsEtsStarts\n",
    "    for etsStart in allEtsStarts:\n",
    "        \n",
    "        # de novo \n",
    "        if etsStart in denovoEtsstarts:\n",
    "            variantTypeList.append('De Novo Ets')\n",
    "            refAffList.append('')\n",
    "            altAffList.append('')\n",
    "            fcList.append('')\n",
    "            ref8merlist.append('')\n",
    "            alt8merlist.append('')\n",
    "            effectList.append(effect)\n",
    "            pvalueList.append(pval)\n",
    "            groupList.append(group)\n",
    "            enList.append(idx)\n",
    "            seqTestedList.append(seqalt)\n",
    "            \n",
    "        # deleted\n",
    "        elif etsStart in deletedetsStarts:\n",
    "            variantTypeList.append('Deleted Ets')\n",
    "            refAffList.append('')\n",
    "            altAffList.append('')\n",
    "            fcList.append('')\n",
    "            ref8merlist.append('')\n",
    "            alt8merlist.append('')\n",
    "            effectList.append(effect)\n",
    "            pvalueList.append(pval)\n",
    "            groupList.append(group)\n",
    "            enList.append(idx)\n",
    "            seqTestedList.append(seqalt)\n",
    "            \n",
    "        # decreased, increased\n",
    "        else:\n",
    "            \n",
    "            ref8mer=zrs_seq[etsStart:etsStart+8]\n",
    "            alt8mer=seqalt [etsStart:etsStart+8]\n",
    "            \n",
    "            refaff=Seq2EtsAff[ref8mer]\n",
    "            altaff=Seq2EtsAff[alt8mer]\n",
    "            fc=altaff/refaff\n",
    "            \n",
    "            if fc<=1: variantTypeList.append('Decrease Ets Aff')\n",
    "            else:     variantTypeList.append('Increase Ets Aff')\n",
    "            \n",
    "            refAffList.append(refaff)\n",
    "            altAffList.append(altaff)\n",
    "            fcList.append(fc)\n",
    "            ref8merlist.append(ref8mer)\n",
    "            alt8merlist.append(alt8mer)\n",
    "            effectList.append(effect)\n",
    "            pvalueList.append(pval)\n",
    "            groupList.append(group)\n",
    "            enList.append(idx)\n",
    "            seqTestedList.append(seqalt)\n",
    "           \n",
    "    # iterate over all non ets siter\n",
    "    nonEtsChanges = mismatchSet - altEtsBases - zrsEtsBases\n",
    "    for nonEts in nonEtsChanges:\n",
    "        variantTypeList.append('Not in Ets')\n",
    "        refAffList.append('')\n",
    "        altAffList.append('')\n",
    "        fcList.append('')\n",
    "        ref8merlist.append('')\n",
    "        alt8merlist.append('')\n",
    "        effectList.append(effect)\n",
    "        pvalueList.append(pval)\n",
    "        groupList.append(group)\n",
    "        enList.append(idx)\n",
    "        seqTestedList.append(seqalt)\n",
    "        \n",
    "    # break\n",
    "            \n",
    "c2v={'p-value':pvalueList,\n",
    "'effect':effectList,\n",
    "'variant-type':variantTypeList,\n",
    "'aff-ref':refAffList,\n",
    "'aff-alt':altAffList,\n",
    "'aff-fc':fcList,\n",
    "'kmer-ref':ref8merlist,\n",
    "'kmer-alt':alt8merlist,\n",
    "'group':groupList,\n",
    "'en':enList,\n",
    "'seq-tested':seqTestedList}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2acd5250-8b4c-49c9-a09e-bc3517597044",
   "metadata": {},
   "outputs": [],
   "source": [
    "df=pd.DataFrame(c2v)\n",
    "\n",
    "\n",
    "\n",
    "df['-log(p)']=df['p-value'].apply(lambda p: -np.log10(p) if p<1 else 0)\n",
    "df['-log(p)*direction']=df.apply(lambda r: r['-log(p)'] if r['effect']>=0 else -r['-log(p)'],axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6422b263-8ca8-4f28-b5c0-b05c37ac9681",
   "metadata": {},
   "outputs": [],
   "source": [
    "df['variant-type'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4e1f2b2f-9fc2-45c7-9a22-6356be77a9c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_csv('FOR-NATURE-ALL-DATA-kvon-corrected-pvals.tsv',sep='\\t',index=None)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7a13f136-ed2d-49fc-96ef-c105e6a2fabb",
   "metadata": {},
   "source": [
    "# Begin plotting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "05cf4d09-9557-4c37-bb78-2d6c2c76e36e",
   "metadata": {},
   "outputs": [],
   "source": [
    "plotDF=df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c64f4a70-8ce0-4405-8a03-f152c7231e0e",
   "metadata": {},
   "outputs": [],
   "source": [
    "logSigCutoff=-np.log10(.05)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16175d48-e158-4c08-a592-06d015490f48",
   "metadata": {},
   "outputs": [],
   "source": [
    "gi='2percent'\n",
    "\n",
    "tempDF=plotDF.loc[plotDF['group']==gi,:]\n",
    "\n",
    "############################################################\n",
    "# Get Data\n",
    "############################################################\n",
    "\n",
    "xticklabels=['All Other Snvs','Increase Ets Aff']\n",
    "\n",
    "#           d0           d1            d2                 d3                 d4    \n",
    "dataLabels=['Not in Ets','Deleted Ets','Decrease Ets Aff','Increase Ets Aff','De Novo Ets']\n",
    "d1,d2,d3,d4=[(tempDF['variant-type']==di) for di in dataLabels[1:]]\n",
    "d0=(df['variant-type']==dataLabels[0])\n",
    "\n",
    "D0=df.loc[d0,'-log(p)*direction'].replace([np.inf,-np.inf],np.NaN).dropna().tolist()\n",
    "D1,D2,D3,D4=[tempDF.loc[di,'-log(p)*direction'].replace([np.inf,-np.inf],np.NaN).dropna().tolist() for di in [d1,d2,d3,d4]]\n",
    "\n",
    "data=[D0+D1+D2+D4,D3]\n",
    "\n",
    "############################################################\n",
    "# Plots\n",
    "############################################################\n",
    "\n",
    "fig,ax=plt.subplots(1,figsize=(5,5),dpi=150)\n",
    "\n",
    "plotType='box' # box or violin\n",
    "\n",
    "colors=['lightgrey','red','tomato','lightgreen','green']\n",
    "if plotType=='violin':\n",
    "\n",
    "\n",
    "\n",
    "    sns.violinplot(data=data,ax=ax,color='white',cut=0,width=.5,inner=None,scale='width') \n",
    "\n",
    "    for i,(d,ci) in enumerate(zip(data,colors)):\n",
    "        y=np.mean(d)\n",
    "        print(i,y)\n",
    "        ax.plot([i-.3,i+.3],[y,y],color=ci)\n",
    "\n",
    "for xi,di in enumerate(data):\n",
    "    for lc, yi in enumerate(di):\n",
    "        if lc>1000: break\n",
    "        xjittermax=.2\n",
    "        xijitter=np.random.uniform(-xjittermax,xjittermax)\n",
    "        if yi>=logSigCutoff:            color='lightgreen'\n",
    "        elif yi<=-logSigCutoff:          color='tomato'\n",
    "        else:                           color='lightgrey'\n",
    "\n",
    "        ax.scatter(xijitter+xi,yi, color=color,s=8,zorder=100,alpha=1)#,cut=0) \n",
    "\n",
    "if plotType=='box':\n",
    "\n",
    "    sns.boxplot(data=data,ax=ax,color='white',width=.8,fliersize=0)     \n",
    "\n",
    "ax.set_title(f'Kvon 2020\\nMultiple Mutations Per Measurement\\n{gi}')\n",
    "\n",
    "ax.set_xticklabels([dl.replace(' ','\\n')+f'\\n{len(di)}' for dl,di in zip(xticklabels,data)])\n",
    "ax.set_ylabel('-log(fdr) with MPRA direction')\n",
    "ax.set_xlabel('\\nEffect')\n",
    "\n",
    "ax.spines['top'].set_visible(False)\n",
    "ax.spines['right'].set_visible(False)\n",
    "\n",
    "# ax.set_ylim(-6.5,9)\n",
    "\n",
    "############################################################\n",
    "# Do Stats\n",
    "############################################################\n",
    "\n",
    "_,p=stats.mannwhitneyu(data[0],data[1],alternative='less')\n",
    "\n",
    "print(f'all other vs increase ets:  p={sf.round(p,3)}')\n",
    "\n",
    "############################################################\n",
    "# add indicator of significance\n",
    "############################################################\n",
    "xmin=-.5\n",
    "xmax=1.5\n",
    "\n",
    "ax.axhline(logSigCutoff,color='green',ls='--',alpha=.4)\n",
    "ax.axhline(-logSigCutoff,color='red',ls='--',alpha=.4)\n",
    "\n",
    "ax.set_xlim(xmin,xmax)\n",
    "\n",
    "plt.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
